---
title: "OB capture enrichment"
author: "kanazian"
date: "`r Sys.Date()`"
output: github_document
always_allow_html: true
---

#Goal
Determine how consistent enrichment is between biological and technical replicates of mouse OBs. Determine the range of enrichment fold for all ORs.

# setup and load data
```{r setup, message=F}
library(tidyverse)
library(heatmaply)

rawdat <- read_csv("~/Desktop/obmap/r_analysis/data/capture_enrichment/obmap_enrichmenteffect_gen25.csv") %>%
  filter(!is.na(gene_name)) %>%
  mutate(isOR = str_detect(gene_name, "^Olfr"),
         isPS = str_detect(gene_name, "-ps"),
         isTaar = str_detect(gene_name, "^Taar"),
         isORTaar = as.logical(isOR + isTaar),
         isVmn = str_detect(gene_name, "^Vmn"),
         isFpr = str_detect(gene_name, "^Fpr"),
         isTarget = case_when(gene_name == "Omp" ~ T,
                              gene_name == "Taf1b" ~ T,
                              str_detect(gene_name, "Olfr") ~ T,
                              str_detect(gene_name, "Taar") ~ T,
                              TRUE ~ F))

probestats <- read_tsv("~/Desktop/obmap/r_analysis/data/capture_enrichment/kzSC1_RNA_coverage.txt") %>% 
  rename ("gene_name" = "REGION_NAME") %>% janitor::clean_names()

#need to remove ectopic ORs
ectopic <- c("Olfr287","Olfr32", "Olfr1033", "Olfr361")
olfrs <- rawdat %>% 
  filter(isOR == T) %>% 
  filter(isPS == F) %>% 
  filter(!(gene_name %in% ectopic)) %>% 
  left_join(probestats, by = "gene_name")

taars <- rawdat %>% 
  filter(isTaar == T) %>% 
  filter(isPS == F) %>% 
  filter(!(gene_name %in% ectopic)) %>% 
  left_join(probestats, by = "gene_name")

olfrstaars <- rawdat %>% 
  filter(isORTaar == 1) %>% 
  filter(isPS == F) %>% 
  filter(!(gene_name %in% ectopic)) %>% 
  left_join(probestats, by = "gene_name")

targets <- rawdat %>% 
  left_join(probestats, by = "gene_name") %>% 
  mutate(isProbed = ifelse(is.na(frac_probe_coverage), F, T)) %>% 
  mutate(doubletarget = as_factor(isTarget*isProbed))
```


# check fold enrichment M6
```{r}
pseudod <- targets %>% 
  mutate(nocap_ps = M6nocapOB + 0.01,
         cap_ps = M6OB + 0.01)

#FIG1A
m6ag_ortaar <- ggplot() +
  geom_point(data = pseudod %>% filter(isORTaar == 0), 
             aes(nocap_ps, cap_ps), 
             color = "#010000", alpha = 0.5) +
  geom_point(data = pseudod %>% filter(isORTaar == 1), 
             aes(nocap_ps, cap_ps), 
             color = "#ff0000", alpha = 0.5) +
  labs(x = "Pre-capture expression (TPM)", y = "Post-capture expression (TPM)") +
  theme_bw() +
  scale_x_log10(labels = scales::trans_format("log10", scales::math_format(10^.x))) +
  scale_y_log10(labels = scales::trans_format("log10", scales::math_format(10^.x))) +
  annotation_logticks(side = "bl", outside = F) +
  theme_kz()

m6ag_vmn <- ggplot() +
  geom_point(data = pseudod %>% filter(isVmn == 0), 
             aes(nocap_ps, cap_ps), 
             color = "#010000", alpha = 0.5) +
  geom_point(data = pseudod %>% filter(isVmn == 1), 
             aes(nocap_ps, cap_ps), 
             color = "#ff0000", alpha = 0.5) +
  labs(x = "Pre-capture expression (TPM)", y = "Post-capture expression (TPM)") +
  theme_bw() +
  scale_x_log10(labels = scales::trans_format("log10", scales::math_format(10^.x))) +
  scale_y_log10(labels = scales::trans_format("log10", scales::math_format(10^.x))) +
  annotation_logticks(side = "bl", outside = F) +
  theme_kz()

m6ag_probed <- ggplot() +
  geom_point(data = pseudod %>% filter(isProbed == 0), 
             aes(nocap_ps, cap_ps), 
             color = "#010000", alpha = 0.5) +
  geom_point(data = pseudod %>% filter(isProbed == 1), 
             aes(nocap_ps, cap_ps), 
             color = "#ff0000", alpha = 0.5) +
  labs(x = "Pre-capture expression (TPM)", y = "Post-capture expression (TPM)") +
  theme_bw() +
  scale_x_log10(labels = scales::trans_format("log10", scales::math_format(10^.x))) +
  scale_y_log10(labels = scales::trans_format("log10", scales::math_format(10^.x))) +
  annotation_logticks(side = "bl", outside = F) +
  theme_kz()

m6ag_legend <- ggplot() +
  geom_point(data = pseudod %>% filter(isORTaar == 0), 
             aes(nocap_ps, cap_ps, color = "Non-OR/Taar"), alpha = 0.5) +
  geom_point(data = pseudod %>% filter(isORTaar == 1), 
             aes(nocap_ps, cap_ps, color = "OR/Taar"), alpha = 0.5) +
  scale_color_manual(values = c("#010000", "#ff0000")) + 
  labs(x = "Pre-capture expression (TPM)", y = "Post-capture expression (TPM)") +
  theme_bw() +
  scale_x_log10(labels = scales::trans_format("log10", scales::math_format(10^.x))) +
  scale_y_log10(labels = scales::trans_format("log10", scales::math_format(10^.x))) +
  annotation_logticks(side = "bl", outside = F) +
  theme_kz()
```


```{r}
print("M6 stats")
#number of ORs with reads
sum(length(which(olfrstaars$M6nocapOB > 0)))
sum(length(which(olfrstaars$M6OB > 0)))

mean(olfrstaars$M6nocapOB)
mean(olfrstaars$M6OB)

median(olfrstaars$M6nocapOB)
median(olfrstaars$M6OB)

m6ag <- targets %>% 
  filter(isPS == F) %>%
  mutate(checkany = M6nocapOB * M6OB,
         cap_ps = M6OB + 0.01,
         nocap_ps = M6nocapOB + 0.01,
         checkboth = M6nocapOB + M6OB,
         foldenr = cap_ps/nocap_ps,
         gene_family = case_when(isOR == T ~ "OR",
                                 isTaar == T ~ "TAAR",
                                 isVmn == T ~ "VMNR",
                                 isFpr == T ~ "FPR",
                                 isProbed == T ~ "Marker"))

m6ag_clean <- m6ag %>% filter(checkany > 0)

m6ot <- m6ag %>%
  filter(gene_family %in% c("OR", "TAAR"))

m6ot_clean <- m6ag_clean %>%
  filter(gene_family %in% c("OR", "TAAR"))

cor(m6ot_clean$M6OB, m6ot_clean$M6nocapOB, method = "s")
cor(olfrstaars$M6OB, olfrstaars$M6nocapOB, method = "spearman")


#SUPFIG
m6ortaaronly <- m6ot %>% 
  filter(gene_family %in% c("OR", "TAAR")) %>%
  ggplot() +
  geom_point(aes(nocap_ps, cap_ps, color = gene_family)) +
  geom_smooth(data = m6ot_clean, aes(nocap_ps, cap_ps), method = "lm", se=F) +
  scale_color_manual(values = c("#010000", "#ff0000")) + 
  labs(x = "Pre-capture Expression (TPM)", y = "Post-capture Expression (TPM)") +
  theme_bw() +
  scale_y_log10(labels = scales::trans_format("log10", scales::math_format(10^.x))) +
  scale_x_log10(labels = scales::trans_format("log10", scales::math_format(10^.x))) +
  annotation_logticks(sides = "bl", outside = F) +
  theme_kz() +
  theme(legend.position = "none")

m6genefamsolid <- m6ag %>% 
  filter(!is.na(gene_family)) %>%
  ggplot() +
  geom_point(aes(nocap_ps, cap_ps, color = gene_family)) +
  scale_color_manual(values = c("#FEC409", "#69747B", "#BE0004", "#73B404", "#4BC2D1")) +
  labs(x = "Pre-capture Expression (TPM)", y = "Post-capture Expression (TPM)") +
  theme_bw() +
  scale_y_log10(labels = scales::trans_format("log10", scales::math_format(10^.x))) +
  scale_x_log10(labels = scales::trans_format("log10", scales::math_format(10^.x))) +
  annotation_logticks(sides = "bl", outside = F) +
  theme_kz() +
  theme(legend.position = "none")

m6genefamalpha <- m6ag %>% 
  filter(!is.na(gene_family)) %>%
  ggplot() +
  geom_point(aes(nocap_ps, cap_ps, color = gene_family), alpha = 0.6) +
  scale_color_manual(values = c("#FEC409", "#69747B", "#BE0004", "#73B404", "#4BC2D1")) +
  labs(x = "Pre-capture Expression (TPM)", y = "Post-capture Expression (TPM)") +
  theme_bw() +
  scale_y_log10(labels = scales::trans_format("log10", scales::math_format(10^.x))) +
  scale_x_log10(labels = scales::trans_format("log10", scales::math_format(10^.x))) +
  annotation_logticks(sides = "bl", outside = F) +
  theme_kz() +
  theme(legend.position = "none")


secm6ag <- targets %>% 
  select(gene_name, contains("M6nocap"), contains("M6S"), contains("is")) %>%
  filter(isPS == F) %>%
  mutate(M6nocap = M6nocap0102 + M6nocap03 + M6nocap04 + M6nocap05 + M6nocap06 + 
           M6nocap07 + M6nocap08 + M6nocap09 + M6nocap10 + M6nocap11 + 
           M6nocap12 + M6nocap13 + M6nocap14 + M6nocap15 + M6nocap16 + 
           M6nocap17 + M6nocap18 + M6nocap19 + M6nocap20 + M6nocap21 + 
           M6nocap22 + M6nocap23 + M6nocap24,
         M6cap = M6S01and02 + M6S03 + M6S04 + M6S05 + M6S06 + M6S07 + M6S08 +
           M6S09 + M6S10 + M6S11 + M6S12 + M6S13 + M6S14 + M6S14 + M6S15 + M6S16 +
           M6S17 + M6S18 + M6S19 + M6S20 + M6S21 + M6S22 + M6S23 + M6S24,
         M6enr = (M6cap + 1) / (M6nocap + 1),
         cap_ps = M6cap + 0.01,
         nocap_ps = M6nocap + 0.01,
         checkany = M6nocap * M6cap,
         checkboth = M6nocap + M6cap,
         gene_family = case_when(isOR == T ~ "OR",
                                 isTaar == T ~ "TAAR",
                                 isVmn == T ~ "VMNR",
                                 isFpr == T ~ "FPR",
                                 isProbed == T ~ "Marker"))
cor(secm6ag$M6nocap, secm6ag$M6cap, method = "s")

secm6ag_clean <- secm6ag %>% filter(checkany > 0)

secm6ot <- secm6ag %>%
  filter(gene_family %in% c("OR", "TAAR"))

secm6ot_clean <- secm6ag_clean %>%
  filter(gene_family %in% c("OR", "TAAR"))

m6ss_ag <- secm6ag %>%
  ggplot() +
  geom_point(data = secm6ag, aes(nocap_ps, cap_ps)) +
  labs(x = "Pre-capture Expression (TPM)", 
       y = "Post-capture Expression (TPM)") +
  theme_bw() +
  scale_y_log10(labels = scales::trans_format("log10", scales::math_format(10^.x))) +
  scale_x_log10(labels = scales::trans_format("log10", scales::math_format(10^.x))) +
  annotation_logticks(sides = "bl", outside = F) +
  theme_kz() +
  theme(legend.position = "none")

m6ss_ortaar <- secm6ot %>%
  filter(gene_family %in% c("OR", "TAAR")) %>%
  ggplot() +
  geom_point(aes(nocap_ps, cap_ps, color = gene_family)) +
  geom_smooth(data = secm6ot_clean, aes(nocap_ps, cap_ps), method = "lm", se=F) +
  scale_color_manual(values = c("#010000", "#ff0000")) + 
  labs(x = "Pre-capture Expression (TPM)", 
       y = "Post-capture Expression (TPM)") +
  theme_bw() +
  scale_y_log10(labels = scales::trans_format("log10", scales::math_format(10^.x))) +
  scale_x_log10(labels = scales::trans_format("log10", scales::math_format(10^.x))) +
  annotation_logticks(sides = "bl", outside = F) +
  theme_kz() +
  theme(legend.position = "none")

m6ss_genefamsolid <- secm6ag %>% 
  filter(!is.na(gene_family)) %>%
  ggplot() +
  geom_point(aes(nocap_ps, cap_ps, color = gene_family)) +
  scale_color_manual(values = c("#FEC409", "#69747B", "#BE0004", "#73B404", "#4BC2D1")) +
  labs(x = "Pre-capture Expression (TPM)", y = "Post-capture Expression (TPM)") +
  theme_bw() +
  scale_y_log10(labels = scales::trans_format("log10", scales::math_format(10^.x))) +
  scale_x_log10(labels = scales::trans_format("log10", scales::math_format(10^.x))) +
  annotation_logticks(sides = "bl", outside = F) +
  theme_kz() +
  theme(legend.position = "none")

m6ss_genefamalpha <- secm6ag %>% 
  filter(!is.na(gene_family)) %>%
  ggplot() +
  geom_point(aes(nocap_ps, cap_ps, color = gene_family), alpha = 0.6) +
  scale_color_manual(values = c("#FEC409", "#69747B", "#BE0004", "#73B404", "#4BC2D1")) +
  labs(x = "Pre-capture Expression (TPM)", y = "Post-capture Expression (TPM)") +
  theme_bw() +
  scale_y_log10(labels = scales::trans_format("log10", scales::math_format(10^.x))) +
  scale_x_log10(labels = scales::trans_format("log10", scales::math_format(10^.x))) +
  annotation_logticks(sides = "bl", outside = F) +
  theme_kz() +
  theme(legend.position = "none")

#FIG1C
#844 in Post, 412 in Pre
m6slide <- olfrs %>% 
  select(gene_name, M6OB, M6nocapOB) %>%
  mutate(check00 = M6OB + M6nocapOB) %>%
  filter(check00 > 0) %>%
  rename("Post-capture" = "M6OB",
         "Pre-capture" = "M6nocapOB") %>%
  gather(key = "type", value = "val", `Post-capture`, `Pre-capture`) %>%
  filter(val > 0) %>%
  arrange(desc(type)) %>% 
  mutate(type_fct = as_factor(type)) %>%
  select(-type) %>% rename("type" = "type_fct") %>%
  ggplot() + 
  geom_density(aes(val, color = type), size = 2) + 
  scale_color_manual(values = c("#010000", "#ff0000")) +
  labs(x = "Expression (TPM)", 
       y = "OR genes") +
  scale_x_log10(labels = scales::trans_format("log10", scales::math_format(10^.x))) + 
  annotation_logticks(sides = "b", outside = F) +
  theme_bw() +
  theme_kz() +
  theme(axis.text.y = element_blank(),
        axis.ticks.y = element_blank())

m6slide_clean <- m6slide + theme(legend.position = "none")

#how many precap
olfrs %>% select(gene_name, M6OB, M6nocapOB) %>% mutate(check00 = M6OB + M6nocapOB) %>% filter(check00 > 0) %>% filter(M6nocapOB > 0) %>% nrow()
#how many postcap
olfrs %>% select(gene_name, M6OB, M6nocapOB) %>% mutate(check00 = M6OB + M6nocapOB) %>% filter(check00 > 0) %>% filter(M6OB > 0) %>% nrow()
```


# biological replicates
```{r}
bioreps <- olfrs %>% 
  select(M4OB, M14XOB, M13XOB, M12OB, 
         M16XOB, M9OBEC, M8OB, M10ECOB) 

#pairwise spearman correlation matrix
bio_pmtx <- matrix(data=0, nrow=ncol(bioreps), ncol=ncol(bioreps))
colnames(bio_pmtx) <- colnames(bioreps)
rownames(bio_pmtx) <- colnames(bioreps)
for (i in 1:ncol(bioreps)) {
  correlation <- vector("numeric", length = ncol(bioreps))
  for (j in 1:ncol(bioreps)) {
    if (i == j) {
      bio_pmtx[i,j] <- 1
    } else {
      bio_pmtx[i,j] <- cor(bioreps[i], bioreps[j], method = "s")
    }
  }
}

#calc mean cor for pairwise 
r1 <- sum(bio_pmtx[1,2:8])
r2 <- sum(bio_pmtx[1,3:8])
r3 <- sum(bio_pmtx[1,4:8])
r4 <- sum(bio_pmtx[1,5:8])
r5 <- sum(bio_pmtx[1,6:8])
r6 <- sum(bio_pmtx[1,7:8])
r7 <- sum(bio_pmtx[1,8])
sum(r1,r2,r3,r4,r5,r6,r7)/sum(1:7)

heatmaply_cor(bio_pmtx)

a <- cor(bioreps$M13XOB, bioreps$M14XOB, method = "s")
b <- cor(bioreps$M13XOB, bioreps$M16XOB, method = "s")
c <- cor(bioreps$M14XOB, bioreps$M16XOB, method = "s")

biopc <- bioreps %>%
  select(M16XOB, M14XOB, M13XOB) %>%
  mutate(M16pc = M16XOB + 1,
         M14pc = M14XOB + 1,
         M13pc = M13XOB + 1,
         checkany = M16XOB * M14XOB * M13XOB,
         check3 = M16XOB + M14XOB + M13XOB)

bio1314 <- ggplot(biopc, aes(M13pc, M14pc)) +
  geom_point(alpha = 0.5) +
  labs(x = "OB1 Expression (TPM)", 
       y = "OB2 Expression (TPM)") +
  theme_bw() +
  scale_y_log10(labels = scales::trans_format("log10", scales::math_format(10^.x))) +
  scale_x_log10(labels = scales::trans_format("log10", scales::math_format(10^.x))) +
  annotation_logticks(sides = "bl", outside = F) +
  theme_kz()

bio1316 <- ggplot(biopc, aes(M13pc, M16pc)) +
  geom_point(alpha = 0.5) +
  labs(x = "OB1 Expression (TPM)", 
       y = "OB3 Expression (TPM)") +
  theme_bw() +
  scale_y_log10(labels = scales::trans_format("log10", scales::math_format(10^.x))) +
  scale_x_log10(labels = scales::trans_format("log10", scales::math_format(10^.x))) +
  annotation_logticks(sides = "bl", outside = F) +
  theme_kz()

bio1416 <- ggplot(biopc, aes(M14pc, M16pc)) +
  geom_point(alpha = 0.5) +
  labs(x = "OB2 Expression (TPM)", 
       y = "OB3 Expression (TPM)") +
  theme_bw() +
  scale_y_log10(labels = scales::trans_format("log10", scales::math_format(10^.x))) +
  scale_x_log10(labels = scales::trans_format("log10", scales::math_format(10^.x))) +
  annotation_logticks(sides = "bl", outside = F) +
  theme_kz()
```


# technical replicates of capture enrichment (2 different arrays on same cDNA, same library)
```{r}
techreps <- olfrs %>% select(NC_10Ar2, NC_10Br2, NC_4Ar2, NC_4Br2, OC_10Ar2, OC_10Br2, OC_4Ar2, OC_4Br2)

one <- cor(techreps$NC_10Ar2, techreps$OC_10Ar2, method = "spearman")
two <- cor(techreps$NC_10Br2, techreps$OC_10Br2, method = "spearman")
three <- cor(techreps$NC_4Ar2, techreps$OC_4Ar2, method = "spearman")
four <- cor(techreps$NC_4Br2, techreps$OC_4Br2, method = "spearman")
mean(c(one, two, three, four))

#pairwise spearman correlation matrix
tech_pmtx <- matrix(data=0, nrow=ncol(techreps), ncol=ncol(techreps))
colnames(tech_pmtx) <- colnames(techreps)
rownames(tech_pmtx) <- colnames(techreps)
for (i in 1:ncol(techreps)) {
  correlation <- vector("numeric", length = ncol(techreps))
  for (j in 1:ncol(techreps)) {
    if (i == j) {
      tech_pmtx[i,j] <- 1
    } else {
    tech_pmtx[i,j] <- cor(techreps[i], techreps[j], method = "spearman")
    }
  }
}

heatmaply_cor(tech_pmtx)

techpc <- techreps %>%
  mutate(oneA = NC_10Ar2 + 1,
         oneB = OC_10Ar2 + 1,
         twoA = NC_10Br2 + 1,
         twoB = OC_10Br2 + 1,
         threeA = NC_4Ar2 + 1,
         threeB = OC_4Ar2 + 1,
         fourA = NC_4Br2 + 1,
         fourB = OC_4Br2 + 1)

techone <- ggplot(techpc, aes(oneB, oneA)) +
  geom_point(alpha = 0.5) +
  labs(x = "OB5 ERCCA Capture1 Expression (TPM)", 
       y = "OB5 ERCCA Capture2 Expression (TPM)") +
  theme_bw() +
  scale_y_log10(labels = scales::trans_format("log10", scales::math_format(10^.x))) +
  scale_x_log10(labels = scales::trans_format("log10", scales::math_format(10^.x))) +
  annotation_logticks(sides = "bl", outside = F) +
  theme_kz()

techtwo <- ggplot(techpc, aes(twoB, twoA)) +
  geom_point(alpha = 0.5) +
  labs(x = "OB5 ERCCB Capture1 Expression (TPM)", 
       y = "OB5 ERCCB Capture2 Expression (TPM)") +
  theme_bw() +
  scale_y_log10(labels = scales::trans_format("log10", scales::math_format(10^.x))) +
  scale_x_log10(labels = scales::trans_format("log10", scales::math_format(10^.x))) +
  annotation_logticks(sides = "bl", outside = F) +
  theme_kz()

techthree <- ggplot(techpc, aes(threeB, threeA)) +
  geom_point(alpha = 0.5) +
  labs(x = "OB4 ERCCA Capture1 Expression (TPM)", 
       y = "OB4 ERCCA Capture2 Expression (TPM)") +
  theme_bw() +
  scale_y_log10(labels = scales::trans_format("log10", scales::math_format(10^.x))) +
  scale_x_log10(labels = scales::trans_format("log10", scales::math_format(10^.x))) +
  annotation_logticks(sides = "bl", outside = F) +
  theme_kz()

techfour <- ggplot(techpc, aes(fourB, fourA)) +
  geom_point(alpha = 0.5) +
  labs(x = "OB4 ERCCB Capture1 Expression (TPM)", 
       y = "OB4 ERCCB Capture2 Expression (TPM)") +
  theme_bw() +
  scale_y_log10(labels = scales::trans_format("log10", scales::math_format(10^.x))) +
  scale_x_log10(labels = scales::trans_format("log10", scales::math_format(10^.x))) +
  annotation_logticks(sides = "bl", outside = F) +
  theme_kz()
```


# ma plots
```{r, message=F}
library(edgeR)

#best technical replicate - all genes
maPlot(rawdat$NC_10Ar2, rawdat$OC_10Ar2)

#best technical replicate - just ORs
maPlot(olfrs$NC_10Ar2, olfrs$OC_10Ar2)

#best biological replicate - just ORs
maPlot(olfrs$M4OB, olfrs$M14XOB)
```

#table 1
```{r}
capeffect_table <- rawdat %>% 
  left_join(probestats, by = "gene_name") %>%
  select(gene_name, ens_raw, M6nocapOB, M6OB, M13XOB, M14XOB, M16XOB, 
         NC_10Ar2, NC_10Br2, NC_4Ar2, NC_4Br2, 
         OC_10Ar2, OC_10Br2, OC_4Ar2, OC_4Br2, 
         contains("M6"), isOR, isTaar, isORTaar, isPS, 
         chromosome:bases_w_no_est_cov_due_to_repeats) %>%
  mutate(fold_enrichment = M6OB/M6nocapOB,
         nocapsliceOB = M6nocap0102 + M6nocap03 + M6nocap04 + M6nocap05 + M6nocap06 + 
           M6nocap07 + M6nocap08 + M6nocap09 + M6nocap10 + M6nocap11 + 
           M6nocap12 + M6nocap13 + M6nocap14 + M6nocap15 + M6nocap16 + 
           M6nocap17 + M6nocap18 + M6nocap19 + M6nocap20 + M6nocap21 + 
           M6nocap22 + M6nocap23 + M6nocap24,
         capsliceOB = M6S01and02 + M6S03 + M6S04 + M6S05 + M6S06 + M6S07 + M6S08 +
           M6S09 + M6S10 + M6S11 + M6S12 + M6S13 + M6S14 + M6S14 + M6S15 + M6S16 +
           M6S17 + M6S18 + M6S19 + M6S20 + M6S21 + M6S22 + M6S23 + M6S24,
         slice_fold_enrichment = capsliceOB/nocapsliceOB) %>% 
  rename("gene" = "gene_name", "ens_gene" = "ens_raw", "capOB" = "M6OB", "nocapOB" = "M6nocapOB", 
         "OB1" = "M13XOB", "OB2" = "M14XOB", "OB3" = "M16XOB", 
         "OB4_A_Cap1" = "OC_4Ar2", "OB4_B_Cap1" = "OC_4Br2", 
         "OB5_A_Cap1" = "OC_10Ar2", "OB5_B_Cap1" = "OC_10Br2", 
         "OB4_A_Cap2" = "NC_4Ar2", "OB4_B_Cap2" = "NC_4Br2", 
         "OB5_A_Cap2" = "NC_10Ar2", "OB5_B_Cap2" = "NC_10Br2") %>%
  select(gene, ens_gene, capOB, nocapOB, fold_enrichment, 
         OB1, OB2, OB3, 
         OB4_A_Cap1, OB4_A_Cap2, OB4_B_Cap1, OB4_B_Cap2, 
         OB5_A_Cap1, OB5_A_Cap2, OB5_B_Cap1, OB5_B_Cap2, 
         capsliceOB, nocapsliceOB, slice_fold_enrichment, everything()) %>%
  arrange(desc(isORTaar))

write_csv(capeffect_table, "~/Desktop/capeffect/capeffect_table_210422.csv")


SaveSquarePlots(techone, techtwo, techthree, techfour, bio1314, bio1316, bio1416, m6slide, m6allgenes0.01nops, m6allgenes0.01ps, m6ag0.01nopslegend, m6ag0.01pslegend, m6ortaar_sectionsums, m6ortaaronly, path = "~/Desktop/capeffect/")
```


#use m6nocap and m6cap samples (1-24, OB) to determine which ORs dropout 
```{r}
m6samples <- olfrstaars %>% select(gene_name, contains("M6"))

m6nocap <- m6samples %>% select(gene_name, contains("nocap"))
m6nocapSum <- m6nocap %>% 
  select(-gene_name, -M6nocapOB) %>% 
  mutate(noCapSum = rowSums(.)) %>% 
  pull(noCapSum) 

m6cap <- m6samples %>% 
  select(gene_name, !contains("nocap")) %>% 
  select(gene_name, contains("S"), M6OB)
m6capSum <- m6cap %>% 
  select(-gene_name, -M6OB) %>% 
  mutate(capSum = rowSums(.)) %>% 
  pull(capSum)

m6compare <- tibble(gene = m6samples$gene_name,
                    noCap = m6nocapSum,
                    Cap = m6capSum,
                    noCapOB = m6samples$M6nocapOB,
                    CapOB = m6samples$M6OB) %>%
  mutate(secComp = ifelse(Cap > noCap, 
                          100, ifelse(Cap == noCap,
                                    10, 0)),
         obComp = ifelse(CapOB > noCapOB, 
                         100, ifelse(CapOB == noCapOB,
                                   10, 0)),
         accord = secComp + obComp,
         status = ifelse(accord == 200, "Enriched", 
                    ifelse(accord == 110, "Likely_enriched", 
                      ifelse(accord == 100, "Variable", 
                        ifelse(accord == 20, "Even", 
                          ifelse(accord < 20, "Dropout", NA))))),
         enriched = str_detect(tolower(status), "enriched"))

notEnriched <- m6compare %>% filter(enriched == F)
notEnrORs <- notEnriched %>% pull(gene)
paste0("There are ", length(notEnrORs), " non-enriched ORs")

ggplot(notEnriched) +
  geom_point(aes(noCap, Cap, color = enriched))

ggplot(notEnriched) +
  geom_point(aes(noCapOB, CapOB, color = enriched))
```


# fold enrichment using genomic DNA controls
b6gen samples not included in the updated set of gencode v25 samples due to being 25bp PE
```{r, eval=F}
b6dna <- olfrs %>% select(gene_id, b6cap100, b6nocap7, b6nocap9, m4cap, m4uncap, M6OB, M6nocapOB) %>% mutate(foldenr7 = b6cap100/(b6nocap7 + 0.1), foldenrm4 = m4cap/(m4uncap + 0.1), foldenrm6 = M6OB/(M6nocapOB + 0.1)) %>% arrange(desc(foldenrm6)) %>% mutate(order = factor(gene_id, gene_id))

ggplot(b6dna, aes(b6nocap7, b6cap100)) +
  geom_point()

ggplot(b6dna, aes(m4uncap, m4cap)) +
  geom_point()

ggplot(b6dna, aes(M6nocapOB, M6OB)) +
  geom_point()

ggplot(b6dna, aes(order, foldenrm6)) +
  geom_bar(stat="identity") +
  ggtitle("M6OB RNA fold enrichment")

ggplot(b6dna, aes(foldenrm4, foldenrm6)) + 
  geom_point() +
  geom_smooth(method = "lm")

ggplot(b6dna, aes(m4cap, M6OB)) + 
  geom_point() +
  geom_smooth(method = "lm")

ggplot(b6dna, aes(m4uncap, M6nocapOB)) + 
  geom_point() +
  geom_smooth(method = "lm")

```


# M4 but avoid since Miseq 25bp
```{r}
olfrs %>% ggplot() +
  geom_point(aes(M4OBnocap, M4OBcap))

#number of ORs with reads
print("M4 stats")
sum(length(which(olfrs$M4OBnocap > 0)))
sum(length(which(olfrs$M4OBcap > 0)))

mean(olfrs$M4OBnocap)
mean(olfrs$M4OB)

median(olfrs$M4OBnocap)
median(olfrs$M4OB)

m4clean <- olfrs %>% 
  select(gene_name, M4OBnocap, M4OB) %>%
  mutate(checkany = M4OBnocap * M4OB,
         checkboth = M4OBnocap + M4OB,
         foldenr = M4OB/M4OBnocap) %>%
  filter(checkboth > 0)

cor(m4clean$M4OB, m4clean$M4OBnocap, method = "spearman")
cor(olfrs$M4OB, olfrs$M4OBnocap, method = "spearman")

prepostm4 <- ggplot(m4clean) +
  geom_point(aes(M4OBnocap, M4OB)) +
  geom_smooth(aes(M4OBnocap, M4OB), method = "lm", se=F) +
  labs(x = "Pre-capture Expression (TPM)", y = "Post-capture Expression (TPM)") +
  theme_cowplot()

compare <- tibble(Olfrname = olfrs$gene_name,
                    m6nocap = olfrs$M6nocapOB,
                    m6cap = olfrs$M6OB,
                    m4nocap = olfrs$M4OBnocap,
                    m4cap = olfrs$M4OBcap,
                    probefrac = olfrs$FRAC_PROBE_COVERAGE) %>%
  mutate(m6enr = (m6cap + 1)/(m6nocap + 1),
         m4enr = (m4cap + 1)/(m4nocap + 1))

# more abundant ORs were enriched more
ggplot(compare, aes(m6cap, m6enr)) +
  geom_point()

ggplot(compare, aes(m6enr, probefrac)) + geom_point(alpha = 0.25) + geom_smooth()

ggplot(compare, aes(m4cap, m4enr)) +
  geom_point()

ggplot(compare, aes(m4enr, probefrac)) + geom_point(alpha = 0.25) + geom_smooth()

ggplot(compare, aes(m6enr, m4enr)) + geom_point(alpha = 0.25) + geom_smooth()

#992 in Post, 496 in Pre
m4slide <- olfrs %>% 
  select(gene_name, M4OBcap, M4OBnocap) %>%
  mutate(check00 = M4OBcap + M4OBnocap) %>%
  filter(check00 > 0) %>%
  rename("Post_capture" = "M4OBcap",
         "Pre_capture" = "M4OBnocap") %>%
  gather(key = "type", value = "val", Post_capture, Pre_capture) %>%
  filter(val > 0) %>%
  ggplot() + 
  geom_density(aes(val, color = type)) + 
  labs(x = "Expression (TPM)", 
       y = "OR genes") +
  scale_x_log10(labels = scales::trans_format("log10", scales::math_format(10^.x))) + 
  annotation_logticks(sides = "b", outside = F) +
  theme_cowplot() +
  theme(axis.text.y = element_blank(),
        axis.ticks.y = element_blank(),
        panel.grid.minor = element_blank())

#1039 in Post, 656 in Pre
comboslide <- olfrs %>% 
  select(gene_name, M4OBcap, M4OBnocap, M6OB, M6nocapOB) %>%
  mutate(check00 = M4OBcap + M4OBnocap + M6OB + M6nocapOB,
         Mcap = M4OBcap + M6OB,
         Mnocap = M4OBnocap + M6nocapOB) %>%
  filter(check00 > 0) %>%
  rename("Post_capture" = "Mcap",
         "Pre_capture" = "Mnocap") %>%
  gather(key = "type", value = "val", Post_capture, Pre_capture) %>%
  filter(val > 0) %>%
  ggplot() + 
  geom_density(aes(val, color = type)) + 
  labs(title = "M4+6 Enrichment Effect", 
       x = "Expression (TPM)", 
       y = "OR genes") +
  scale_x_log10(labels = scales::trans_format("log10", scales::math_format(10^.x))) + 
  annotation_logticks(sides = "b", outside = F) +
  theme_bw() +
  theme(axis.text.y = element_blank(),
        axis.ticks.y = element_blank(),
        panel.grid.minor = element_blank())
```
