---
title: "Does pS6IP basal activity relate to AP OB position"
author: "kanazian"
date: "'r Sys.Date()'"
output: html_notebook
---

# Goal
Define sequence positions that contain Grantham property different residues between ORs that target anterior or posterior in the OB, model a binary classifier SVM on these positions.

# Sample Info
Lists of Ant/Post targeting ORs came from my OBmap data. P21 B6 mice OBs were dissected, embedded in agarose and sectioned at 100um on a vibratome. RNA was extracted, cDNA with SmartSeq, libraries with KAPA hyperplus, capture with Nimblegen, reads sequenced on Hiseq 50bp SE rapid or NextSeq 75bp SE high output, and reads aligned with kallisto against our labs reference. TPM values across the sections were used to generate a weighted average, allowing each gene to have an average position across the number of sections.

Will need to make and use different lists - Ventral AP, Dorsal AP, Class 2 AP

# Setup
Load packages, show path, load data, define functions
```{r setup, include=TRUE, warning=FALSE}
library(seqinr)
library(e1071)
library(gtools)
library(ggsci)
library(cowplot)
library(tidyverse)
library(doParallel)

#load data
Prop.OR <- readRDS("~/Desktop/obmap/r_analysis/data/classifyAP/PropOR.RData")
all_positions <- read_csv("~/Desktop/info_210430.csv")
OR_aln <- read.alignment(file = "~/Desktop/obmap/r_analysis/data/classifyAP/mouseORprotein_clean_v17.aln", format = "clustal")
OR_aln$nam <- as.character(lapply(strsplit(OR_aln$nam, split="/"), "[", 1))
tgrantham <- read.csv("~/Desktop/obmap/r_analysis/data/classifyAP/grantham.csv", header = T)
AAProp <- readRDS("~/Desktop/obmap/r_analysis/data/classifyAP/AAprop.RData")
grantham <- tgrantham[,2:ncol(tgrantham)] 
#eliminate first col which is aa letters of pairwise matrix
colnames(grantham)[ncol(grantham)] <- "-"
rownames(grantham) <- colnames(grantham)
aa <- row.names(grantham)

#functions---------------------------------------------------------------------
#distance between residues for groups
per.aa.dist <- function(OR.aln=OR_aln,ind=ind,combo=combo) {
  seq.sub <- NULL
  for (i in ind) {
    seq.sub <- rbind(seq.sub, strsplit(OR.aln$seq[[i]],NULL)[[1]])
  }
  length <- dim(seq.sub)[2]
  height <- dim(seq.sub)[1]
  
  dist <- NULL
  for (i in 1:length) {
    distance <- NULL
    for (a in 1:dim(combo)[1]) {
      distance <- c(distance, grantham[which(aa==seq.sub[combo[a,1],i]),which(aa==seq.sub[combo[a,2],i])])
    }
    dist <- c(dist, mean(distance,na.rm=T))
  }
  return(dist)
}

#view alignment site, real sequence position, and residue identity with surrounding neighbors
print_aa <- function(aln, pos_vec, surround = 0) {
  for (pos in pos_vec) {
    aa <- str_to_upper(substr(aln, pos - surround, pos + surround))
    dashes_upto_pos <- str_count(substr(aln, 1, pos),"-") + 7
    line <- paste0("Aln: ", pos, 
                   "  Position: ", pos - dashes_upto_pos,
                   "  Residue: ", aa, 
                   "  Pval: ", all_sigs[pos], sep="")
    print(line)
  }
}


Make_prop <- function(sites_v, OR_aln = OR_aln) {
  NewProp <- matrix(0, nrow = length(OR_aln$seq), ncol = length(sites_v)*3)
  PropValues <- NULL
  for (seq in OR_aln$seq) {
    for (site in sites_v) {
      residue <- substr(seq, site, site)
      res_c <- AAProp[which(AAProp$AA == residue), 2]
      res_p <- AAProp[which(AAProp$AA == residue), 3]
      res_v <- AAProp[which(AAProp$AA == residue), 4]
      PropValues <- c(PropValues, res_c, res_p, res_v)
    }
    NewProp[which(OR_aln$seq %in% seq),] <- PropValues
    PropValues <- NULL
  }
  rownames(NewProp) <- OR_aln$nam
  newProp_cn <- NULL
  for (site in sites_v) {
    pos_c <- paste(as.character(site), "_c", sep="")
    pos_p <- paste(as.character(site), "_p", sep="")
    pos_v <- paste(as.character(site), "_v", sep="")
    newProp_cn <- c(newProp_cn, pos_c, pos_p, pos_v)
  }
  colnames(NewProp) <- newProp_cn
  NewProp
}

#output SVM AUC
ROC.gen <- function(ans,add=F) {  
  ans <- ans[order(ans[,1],decreasing=T),]
  tp <- fp <- 0
  TP <- FP <-  NULL
  tp.n <- sum(ans[,2]==1)
  fp.n <- sum(ans[,2]==0)
  l <- dim(ans)[1]
  for (i in 1:l) {
    if (ans[i,2]==1) {tp <- tp+1}
    if (ans[i,2]==0) {fp <- fp+1}
    TP <- c(TP,tp/tp.n)
    FP <- c(FP,fp/fp.n)
  }
  auc <- TP[1:(l-1)]%*%(FP[2:l]-FP[1:(l-1)])
#  plot(c(0,FP),c(0,TP),type="l",col="blue",lwd=3,xlab="False positive rate",ylab="True positive rate",frame.plot=F,xlim=c(0,1),ylim=c(0,1))
  if (add==F) {
    plot(c(0,FP),c(0,TP),type="l",col="blue",lwd=3,xlab=NA,ylab=NA,frame.plot=F)
    title(xlab="False positive rate",ylab="True positive rate",cex.lab=1.5)
    segments(0,0,1,1,lwd=3)
    legend("bottomright",c("protein sequence-based classifier","random"),col=c("blue","black"),lwd=c(3,3),lty=c(1,1),bty="n",cex=1)
#   text(0.1,1,paste("AUC:",round(auc,3)))
  }
  else {points(c(0,FP),c(0,TP),type="l",col="green")}
  plot(0:10, type = "n", xaxt="n", yaxt="n", bty="n", xlab = "", ylab = "")
  text(5,5,paste("p=",wilcox.test(ans[,1]~ans[,2],alternative="less")$p.value))
  text(5,4,paste("AUC:",round(auc,3)))
#  plot(sort(ans[,1]),TP,type="l",col="blue",lwd=3,xlab=NA,ylab=NA,frame.plot=F)
#  points(sort(ans[,1]),FP,type="l",col="green")
#  points(sort(ans[,1]),TP/FP,type="l",col="red")
  return (list(auc,FP,TP))
}
```


# Group low (1) and high (2) pS6-IP basal activity ORs
(1) is a low basal activity group which corresponds to anterior targeting OB ORs
(2) is a high basal activity group which corresponds to posterior targeting OB ORs
Excluding Class 1 ORs since they are biased towards anterior OB and have systematic protein sequence differences to C2 ORs
```{r}
newDE_fdr <- all_positions %>% 
  filter(!is.na(newLFC)) %>% 
  filter(newFDR < 0.15) %>% 
  filter(class == 2) %>%
  mutate(LFCsplit = ifelse(newLFC > 0.2, 
                           "high", ifelse(newLFC < -0.2, 
                                          "low", "mid"))) %>% 
  group_by(LFCsplit) %>%
  mutate(split_rank = ifelse(LFCsplit == "high", 
                             min_rank(newLFC), ifelse(LFCsplit == "low", 
                                                      min_rank(desc(newLFC)), NA))) %>%
  ungroup()

newDE_min <- newDE_fdr %>% 
  filter(LFCsplit != "mid") %>% 
  group_by(LFCsplit) %>% 
  summarise(count = n()) %>% 
  pull(count) %>%
  min()

newDE_final <- newDE_fdr %>% filter(split_rank <= newDE_min) %>% arrange(split_rank)

new_high <- newDE_final %>% filter(LFCsplit == "high") %>% pull(gene)
new_low <- newDE_final %>% filter(LFCsplit == "low") %>% pull(gene)

oldDE_fdr <- all_positions %>% 
  filter(!is.na(oldLFC)) %>% 
  filter(oldFDR < 0.15) %>% 
  filter(class == 2) %>%
  mutate(LFCsplit = ifelse(oldLFC > 0.2, 
                           "high", ifelse(oldLFC < -0.2, 
                                          "low", "mid"))) %>% 
  group_by(LFCsplit) %>%
  mutate(split_rank = ifelse(LFCsplit == "high", 
                             min_rank(oldLFC), ifelse(LFCsplit == "low", 
                                                      min_rank(desc(oldLFC)), NA))) %>%
  ungroup()

oldDE_min <- oldDE_fdr %>% 
  filter(LFCsplit != "mid") %>% 
  group_by(LFCsplit) %>% 
  summarise(count = n()) %>% 
  pull(count) %>%
  min()

oldDE_final <- oldDE_fdr %>% filter(split_rank <= oldDE_min) %>% arrange(split_rank)

old_high <- oldDE_final %>% filter(LFCsplit == "high") %>% pull(gene)
old_low <- oldDE_final %>% filter(LFCsplit == "low") %>% pull(gene)

#look at old vs new overlap
paste0("High Basal OR sets have ", round((length(which(old_high %in% new_high))/length(new_high))*100, digits=2), "% overlap")
paste0("Low Basal OR sets have ", round((length(which(old_low %in% new_low))/length(new_low))*100, digits=2), "% overlap")
```

# Group anterior (1) and posterior (2) OB ORs
Using only class 2 ORs here as well
```{r}
filter_positions <- all_positions %>% 
  filter(class == 2) %>%
  filter(filterFail == 0) %>%
  rowwise() %>%
  mutate(DVscore = sum(ifelse(oe_region == "Dorsal", 1, 0),
                       ifelse(tz_vd == "Dorsal", 2, 0), 
                       na.rm=T)) %>%
  ungroup() %>%
  mutate(DVdiscord1 = case_when(tan_zone == "unusual" ~ "Unusual",
                                DVscore >= 2 ~ "Dorsal",
                                DVscore == 0 ~ "Ventral",
                                DVscore == 1 ~ ifelse(vd_wavg < 10.5, 
                                                      "Ventral_vdmp", "Dorsal_vdmp")),
         DVdiscord2 = ifelse(is.na(DVdiscord1), 
                             case_when(is.na(tz_vd) ~ paste(oe_region, "oere", sep="-"),
                                       is.na(oe_region) ~ paste(tz_vd, "tzvd", sep="-")), 
                             DVdiscord1),
         DVdiscord = case_when(str_detect(DVdiscord2, "Dorsal") ~ "Dorsal",
                               str_detect(DVdiscord2, "Ventral") ~ "Ventral",
                               DVdiscord2 == "Unusual" ~ "Unusual"),
         DVcertainty = case_when(DVscore == 2 ~ "High",
                                 DVscore == 0 ~ "High",
                                 str_detect(DVdiscord2, "_vdmp") ~ "Medium",
                                 str_detect(DVdiscord2, "-") ~ "Low")) %>%
  #group_by(DVdiscord) %>%
  mutate(ant_rank = min_rank(ap_wavg),
         post_rank = min_rank(desc(ap_wavg))) %>%
  ungroup()

dorsal_n <- filter_positions %>% filter(DVdiscord == "Dorsal") %>% 
  pull(ant_rank) %>% max(na.rm=T)
ventral_n <- filter_positions %>% filter(DVdiscord == "Ventral") %>% 
  pull(ant_rank) %>% max(na.rm=T)
all_n <- filter_positions %>% pull(ant_rank) %>% max(na.rm=T)
splitp <- 0.4

# AP groups will consist of 40% of DV group size, with middle 20% not used

# hm indicates sort made from classic heatmap set (m4,6,7,10)
dor_hm_ant <- filter_positions %>%
  filter(DVdiscord == "Dorsal") %>%
  filter(ant_rank < dorsal_n * splitp) %>%
  arrange(ant_rank) %>%
  pull(olfrname)

dor_hm_post <- filter_positions %>%
  filter(DVdiscord == "Dorsal") %>%
  filter(post_rank < dorsal_n * splitp) %>%
  arrange(post_rank) %>%
  pull(olfrname)

ven_hm_ant <- filter_positions %>%
  filter(DVdiscord == "Ventral") %>%
  filter(ant_rank < ventral_n * splitp) %>%
  arrange(ant_rank) %>%
  pull(olfrname)

ven_hm_post <- filter_positions %>%
  filter(DVdiscord == "Ventral") %>%
  filter(post_rank < ventral_n * splitp) %>%
  arrange(post_rank) %>%
  pull(olfrname)

#do without grouping filter position by DVdiscord prior to mutate ant/post rank
all_ant <- filter_positions %>%
  filter(ant_rank < all_n * splitp) %>%
  arrange(ant_rank) %>%
  pull(olfrname)

all_post <- filter_positions %>% 
  filter(post_rank < all_n * splitp) %>%
  arrange(post_rank) %>%
  pull(olfrname)
  
```


# cursory examination of LogFC and AP position
```{r}
all_positions %>% filter(filter == "Pass") %>% filter(oldFDR < 0.15) %>% ggplot() + geom_point(aes(aphm_wavg, oldLFC)) + facet_wrap(~ tz_vd)
```


# Determine residues that are significantly different (compared to random sets) in terms of grantham properties
```{r, eval=F}
antGroup <- all_ant
postGroup <- all_post

rawAntIdx <- which(OR_aln$nam %in% antGroup)
rawPostIdx <- which(OR_aln$nam %in% postGroup)
rawAllIdx <- c(rawAntIdx, rawPostIdx)

#AP only alignment, remove middle
AP_seq <- NULL
AP_nam <- NULL
for (i in rawAllIdx) {
  AP_seq <- c(AP_seq, OR_aln$seq[i])
  AP_nam <- c(AP_nam, OR_aln$nam[i])
} #endfor
AP_aln <- list(nam = AP_nam, seq = AP_seq)

#new indexes
AntIdx <- which(AP_aln$nam %in% antGroup)
PostIdx <- which(AP_aln$nam %in% postGroup)
m <- round(length(AntIdx)/4)
combo <- combinations(m,2)

m2 <- length(AntIdx)
combo2 <- combinations(m,2)

AntDistance <- per.aa.dist(OR.aln=AP_aln, ind=AntIdx, combo=combo2)

#Dont make matrix, takes longtime
size <- 1500
kzmtx <- NULL
registerDoParallel(cores=11) 

start <- Sys.time()
kzmtx <- foreach(count=1:size, .combine=rbind) %dopar% {
  ind <- sample(length(AP_aln$nam),m)
  per.aa.dist(OR.aln=AP_aln, ind=ind, combo=combo)
} #end foreach
stop <- Sys.time()
(time <- stop - start)

saveRDS(kzmtx, "~/Desktop/obmap/r_analysis/data/classifyAP/mtx1500_allAP_210515.RDS")
```

##Find significant sites
```{r, eval=F}
#load made mtx
kzmtx <- readRDS("~/Desktop/obmap/r_analysis/data/classifyAP/mtx1500_allAP_210515.RDS")

all_sigs <- NULL
#for each position in alignment, compare distance between RTPind group and allOR group, returned value is the proportion of TRUEs aka RTPind group is less similar than allOR, lower value indicates more similar
for (i in 1:dim(kzmtx)[2]) {
  all_sigs <- c(all_sigs, mean(AntDistance[i]>=kzmtx[,i],na.rm=T)) 
} #end for

All.OR <- NULL
for (i in 1:length(OR_aln$seq)) {
  All.OR <- rbind(All.OR, strsplit(OR_aln$seq[[i]],NULL)[[1]])
} #end for

olfr539aln <- OR_aln$seq[[which(OR_aln$nam == "Olfr539")]]
#consensus seq for mouseorv17
OR_aln$seq[1091][1] <- "MWLCNKTKTWACAEFIPYWRFLFVVVSGKTGFYYVALAGADKQEFCGLCFEMTGQLFILDQHMMMLFLMMMMMMMMEMENNTSSFHPSTTSSSVVTEFILLGLSTDDPELQLLLFVLFLLIYLVTLLGNLLIILLIRLDIHIDSHLLHTPMYFFLSNLSFLDICYSSVTVPKMLWAVNLLSGEEGEKKTISFAGCMTQLFFFHFFGGTECFLLAAMAYDRYVAICKPLHYTTIMSPRVCVLLVLGSWVGGFLHSLIHTLLTFRGNSVVQTALTMRLPFCGSDAGGKNVINHFFCDIPPLLKLASVCSDTSINELVVFVVAGFILLVPFLLILVSYVFILVVTSTILRIPSAEGRRKAFSTCSSHLTVVSLFYGTAIFMYLRPSSSYHNGADVSPDQDKVVSVFYTVVTPMLNPLIYSLRNKDVKGALKKLWILKFRQLLLLKKKKLSSKKKLSGGSRNSLFLRTKGHDKQILLKLDQIGKRALMWCLVSHRSQISDVYYCLQGRASILFKEIGLLTKKNPSKLTEKDLWLPFLHRVEEMIYLQLTCDAVQTQEEGGTV"
OR_aln$nam[1091] <- "consensus"

orcons <- OR_aln$seq[[which(OR_aln$nam == "consensus")]]

length(which(all_sigs < 0.05))
plot(all_sigs, col="red", type="l", ylim=c(0,1)); abline(h=0.05,lty=2)
print_aa(olfr539aln, which(all_sigs < 0.05), surround = 1)

sigs_fdr <- all_sigs
keep.ind <- (colMeans(All.OR=="-") < 0.1)
sigs_fdr[keep.ind] <- p.adjust(sigs_fdr[keep.ind],"fdr")
sigs_fdr[!keep.ind] <- NA

length(which(sigs_fdr < 0.05))
plot(sigs_fdr,type="l",ylim=c(0,1));abline(h=0.05,lty=2)

print_aa(orcons, which(sigs_fdr < 0.05), surround = 1)
which(sigs_fdr < 0.05)
print_aa(olfr539aln, which(sigs_fdr < 0.01), surround = 1)
which(sigs_fdr < 0.01)

allsigs5 <- which(all_sigs < 0.05)
```

#Make lists of significant sites
new for 2021
```{r}
#paper draft dorsal AP
dorAP <- c(129, 130, 132, 144, 153, 155, 156, 158, 162, 163, 164, 165, 167, 169, 170, 171, 172, 175, 176, 177, 179, 180, 183, 184, 185, 190, 191, 193, 195, 197, 198, 199, 203, 204, 208, 210, 212, 214, 216, 220, 222, 223, 224, 229, 230, 237, 239, 241, 242, 248, 249, 251, 257, 270, 271, 273, 275, 277, 279, 280, 282, 284, 285, 295, 303, 304, 306, 307, 308, 309, 310, 311, 312, 313, 328, 330, 332, 334, 337, 339, 345, 350, 351, 356, 357, 358, 360, 361, 362, 363, 367, 368, 371, 374, 382, 384, 388, 393, 394, 395, 397, 398, 399, 400, 401, 402, 404, 410, 418, 419, 421, 422, 424, 427, 428, 430, 433, 434, 435, 436, 438, 440, 449, 452, 453, 455, 456, 457, 462, 463, 464, 466, 468, 471, 472, 480, 485, 486, 488, 489, 493, 494, 496, 500, 503, 508, 511, 512, 513, 515, 517, 518, 519, 521, 530, 532)

dorAP_fdr0.05 <- c(129, 130, 144, 153, 155, 156, 158, 162, 163, 164, 165, 167, 169, 170, 172, 175, 176, 177, 179, 180, 183, 184, 185, 190, 191, 193, 195, 197, 198, 199, 203, 204, 208, 210, 212, 214, 216, 220, 222, 223, 224, 229, 230, 237, 239, 241, 242, 248, 249, 251, 257, 270, 271, 273, 275, 277, 279, 280, 282, 284, 285, 295, 303, 304, 307, 308, 309, 312, 328, 330, 332, 334, 337, 339, 345, 350, 351, 356, 357, 360, 361, 362, 363, 367, 371, 374, 382, 384, 388, 393, 394, 395, 397, 398, 399, 400, 401, 402, 404, 410, 421, 422, 424, 427, 428, 430, 433, 434, 435, 436, 438, 440, 449, 452, 453, 456, 457, 462, 463, 464, 466, 468, 471, 472, 480, 486, 488, 489, 493, 494, 496, 503, 508, 511, 512, 513, 515, 517, 518, 521, 530, 532)

dorAP_fdr0.01 <- c(129, 130, 160, 162, 163, 164, 165, 170, 171, 172, 173, 175, 177, 180, 190, 191, 195, 198, 199, 204, 208, 212, 220, 239, 242, 259, 275, 277, 280, 285, 295, 303, 312, 328, 332, 338, 350, 357, 358, 361, 363, 374, 382, 388, 397, 399, 401, 404, 406, 438, 440, 449, 452, 453, 456, 457, 462, 464, 466, 468, 471, 480, 486, 487, 489, 493, 494, 496, 503, 511, 512, 513, 517, 521, 530)

venAP <- c(31, 39, 129, 130, 149, 158, 160, 161, 162, 163, 164, 165, 167, 170, 171, 172, 174, 175, 184, 190, 193, 195, 198, 199, 203, 204, 208, 216, 220, 228, 229, 231, 237, 240, 242, 259, 267, 268, 271, 274, 275, 277, 280, 301, 306, 307, 311, 312, 332, 350, 357, 358, 361, 362, 382, 388, 390, 394, 396, 397, 399, 401, 404, 406, 414, 416, 417, 419, 423, 427, 442, 444, 452, 455, 456, 457, 459, 462, 463, 466, 471, 480, 485, 486, 487, 489, 494, 500, 503, 511, 514, 517, 530, 532, 533, 561, 594)

venAP_fdr <- c(129, 130, 149, 162, 163, 164, 165, 167, 170, 171, 172, 174, 175, 184, 190, 193, 195, 198, 199, 204, 208, 220, 231, 237, 242, 259, 267, 268, 271, 275, 277, 301, 307, 312, 332, 358, 361, 362, 382, 388, 394, 396, 397, 399, 401, 404, 406, 414, 423, 427, 444, 452, 456, 457, 462, 463, 466, 471, 480, 486, 487, 489, 494, 500, 503, 511, 514, 517, 530, 532)

venAP_fdr0.01 <- c(130, 163, 170, 171, 172, 174, 175, 184, 190, 193, 195, 198, 199, 204, 208, 220, 242, 259, 267, 268, 277, 358, 361, 382, 388, 394, 396, 397, 399, 401, 423, 427, 444, 452, 456, 457, 462, 466, 471, 480, 486, 487, 489, 500, 517, 530)

#kzaln
oldbasal <- c(130, 144, 153, 155, 158, 162, 163, 164, 165, 170, 171, 172, 174, 175, 176, 178, 184, 187, 190, 193, 195, 198, 199, 204, 208, 214, 216, 220, 229, 231, 235, 237, 248, 275, 277, 285, 304, 307, 313, 332, 350, 357, 358, 361, 362, 367, 371, 373, 382, 388, 390, 394, 395, 396, 401, 404, 406, 414, 419, 421, 423, 427, 430, 438, 444, 456, 457, 462, 463, 466, 471, 485, 486, 489, 494, 496, 500, 503, 514, 517, 518, 521, 532, 584)
oldbasal_fdr <- c(130, 162, 163, 165, 170, 171, 184, 187, 193, 195, 198, 208, 231, 235, 275, 304, 313, 332, 357, 358, 388, 394, 396, 401, 406, 414, 421, 423, 427, 438, 456, 457, 462, 466, 471, 486, 489, 500, 503, 514, 517)

newbasal <- c(171, 172, 178, 220, 223, 228, 240, 249, 267, 268, 280, 300, 314, 358, 361, 371, 388, 389, 390, 396, 401, 406, 414, 463, 485, 515, 528, 559)

ventral3d <- c(31, 108, 110, 132, 144, 159, 160, 163, 174, 176, 177, 178, 182, 183, 184, 187, 190, 193, 195, 207, 217, 219, 220, 232, 241, 242, 263, 265, 267, 268, 274, 276, 279, 280, 287, 289, 300, 305, 328, 338, 344, 356, 357, 358, 362, 368, 371, 373, 377, 382, 388, 389, 395, 396, 397, 401, 404, 418, 423, 442, 444, 446, 449, 451, 452, 454, 458, 459, 462, 463, 468, 471, 480, 486, 490, 492, 495, 496, 499, 501, 503, 514, 525, 527, 530, 593)
ventral3d_fdr <- c(159, 163, 177, 178, 183, 187, 193, 207, 217, 220, 232, 263, 265, 267, 268, 274, 276, 279, 280, 289, 300, 305, 338, 344, 356, 358, 371, 373, 377, 382, 388, 395, 396, 401, 404, 418, 423, 442, 444, 446, 452, 458, 462, 471, 486, 492, 495, 499, 514, 525)

dorsal3d <- c(26, 31, 109, 159, 160, 176, 177, 178, 180, 183, 187, 190, 196, 217, 220, 224, 232, 241, 263, 265, 268, 274, 276, 279, 289, 303, 305, 338, 344, 349, 352, 356, 358, 368, 371, 373, 389, 395, 405, 418, 451, 452, 454, 458, 459, 462, 468, 480, 486, 488, 492, 495, 496, 499, 501, 525)
dorsal3d_fdr <- c(178, 217, 344, 356, 358, 371, 418, 501)

ventralhm <- c(26, 31, 39, 108, 109, 110, 149, 159, 163, 174, 176, 177, 178, 180, 182, 183, 184, 187, 193, 207, 217, 219, 220, 224, 232, 240, 241, 263, 265, 267, 268, 274, 276, 279, 280, 289, 300, 305, 307, 315, 328, 331, 338, 344, 356, 358, 362, 368, 371, 373, 377, 382, 387, 388, 389, 395, 396, 397, 401, 404, 418, 423, 446, 451, 452, 453, 454, 455, 458, 459, 462, 463, 466, 471, 480, 486, 488, 490, 492, 495, 496, 499, 503, 511, 514, 525, 530, 593)
ventralhm_fdr <- c(163, 174, 176, 178, 183, 184, 187, 193, 207, 217, 219, 220, 232, 240, 241, 265, 268, 274, 276, 279, 280, 300, 305, 338, 344, 356, 358, 362, 371, 373, 377, 382, 387, 389, 395, 396, 397, 401, 404, 418, 423, 446, 452, 453, 455, 458, 462, 463, 486, 492, 495, 499, 514, 525)

dorsalhm <- c(119, 130, 155, 163, 171, 178, 180, 181, 184, 186, 187, 190, 191, 199, 207, 228, 232, 240, 241, 249, 266, 275, 279, 280, 282, 284, 295, 300, 332, 334, 338, 350, 356, 357, 362, 371, 377, 382, 389, 393, 398, 405, 414, 416, 436, 438, 449, 456, 457, 458, 459, 463, 466, 471, 480, 485, 486, 488, 489, 490, 492, 495, 501, 503, 510, 514, 515, 518, 592)
dorsalhm_fdr <- c(130, 163, 186, 191, 228, 232, 241, 249, 275, 284, 338, 350, 356, 371, 398, 416, 457, 458, 459, 463, 466, 480, 488, 490, 501, 518)
```


```{r, eval=F}
#yue aln
# oldbasal <- c(103, 110, 170, 177, 181, 203, 208, 210, 242, 277, 312, 322, 330, 336, 349, 372, 373, 375, 386, 402, 407, 414, 450)
# 
# newbasal_uncor <- c(106, 108, 110, 124, 135, 137, 156, 168, 177, 187, 210, 212, 243, 253, 256, 277, 330, 331, 347, 371, 402, 414, 443)
# newbasal_adj <- 414
# 
# 
# dorsalhm_uncor <- c(111, 181, 188, 201, 253, 257, 261, 316, 332, 353, 371, 406, 422)
# 
# ventralhm_uncor <- c(103, 116, 123, 131, 178, 196, 199, 203, 213, 217, 226, 229, 233, 253, 261, 277, 280, 281, 295, 296, 312, 316, 319, 321, 323, 328, 330, 343, 349, 356, 357, 367, 371, 382, 386, 397, 400, 405, 406, 408, 422, 439)
# ventralhm_fdr <- c(131, 217, 229, 253, 277, 281, 296, 321, 397)
# 
# 
# dorsal3d_uncor <- c(76, 111, 137, 181, 199, 201, 253, 257, 261, 287, 289, 296, 373)
# 
# ventral3d_uncor <- c(123 ,131 ,169 ,178 ,199 ,222 ,226 ,229 ,233 ,239 ,246 ,253 ,261 ,277 ,281 ,296 ,308, 323 ,326 ,328 ,357 ,368 ,371 ,382 ,386 ,397 ,405 ,406 ,422 ,439)
# ventral3d_adj <- 397
```


#Create matrix of grantham properties (composition, polarity, volume) for each residue in alignment
Composition is ratio of carbon to noncarbon in residue
Polarity is charge of residue
Volume is space taken up by residue 

Making 2 matrixes:
1. based off old positions in Yue's PropOR because it is missing 7 columns (oldPosNewPropOR)
2. based off positions in alignment that have consensus (consPropOR)
```{r}
varProp <- Make_prop(oldbasal_fdr, OR_aln)

#subset ORs
set1 <- which(rownames(varProp) %in% old_low)
set1len <- length(which(rownames(varProp) %in% old_low))
set2 <- which(rownames(varProp) %in% old_high)
set2len <- length(which(rownames(varProp) %in% old_high))

label <- c(rep(1, set1len), rep(0, set2len))
tmp <- rbind(varProp[set1,], varProp[set2,])
tmp_noNA <- tmp[, colSums(is.na(tmp)) == 0]
all.mat <- cbind(label, tmp_noNA)
col.ind <- apply(all.mat, 2, var) != 0
all.mat <- as.data.frame(all.mat[, col.ind])
k <- 1 + log(nrow(varProp))/log(2)

tune(svm, as.factor(label)~., data = all.mat, ranges = list(gamma = 0.0001:2, cost = 0.0001:10))
```

Do SVM
```{r}
data <- all.mat
cost <- 8
gamma <- 1e-04
fold <- 10 
cycle <- 300
test.size = round(dim(data)[1]/fold)
#BIG.TP<-BIG.FP<-AUC<-NULL
ANS <- NULL
set.seed(0)
for (j in 1:cycle) { 
  repeat{
    test.ind <- sample(1:dim(data)[1],test.size)
    test.data <- data[test.ind,]
    if (var(test.data$label)!=0) {break}
  }
  train.data <- data[-test.ind,]
  col.ind <- apply(train.data,2,var)!=0
  col.ind[1] <- TRUE
  train.data <- train.data[,col.ind]
  test.data <- test.data[,col.ind]
  svm.model <- svm(as.factor(label) ~ ., 
                   data = train.data, 
                   kernel="radial", 
                   cost = cost, 
                   gamma = gamma, 
                   scale=T, 
                   probability=TRUE)
  #  svm.model <- svm(as.factor(res) ~ ., data=train.data,kernel="linear",scale=T,probability=TRUE)
  svm.pred <- predict(svm.model, test.data, na.action = na.exclude, probability=TRUE, decision.values = TRUE)
  ans <- cbind(attr(svm.pred, "decision.values"),test.data$label, attr(svm.pred, "probabilities")[,1])
  ANS <- rbind(ANS, ans)
}
ROC.gen(ANS)
text(5,3,paste("cost:",cost))
text(5,2,paste("gamma:",gamma))

#oldbasal = 0.658, 3.68e-114
#oldbasalfdr = 0.602, 8.66e-49
#newbasal = 0.398, 1
#ventral3d = 0.836, 0
#ventral3dfdr = 0.78, 0
#ventralhm = 0.881, 0
#ventralhmfdr = 0.844, 0
#dorsal3d = 0.63, 3.42e-81
#dorsal3dfdr = 0.578, 1.06e-30
#dorsalhm = 0.71, 1.01e-208
#dorsalhmfdr = 0.524, 0.00022
```

Using basal sites to predict Posterior ORs
```{r}
PostORs <- which(rownames(varProp) %in% old_high)
PostORlen <- length(which(rownames(varProp) %in% old_high))
test.lab <- c(rep(1, PostORlen))
test.tmp <- varProp[PostORs,]
test.dat <- cbind(test.lab, test.tmp)
test.ind <- apply(test.dat, 2, var) != 0
test.dat.final <- as.data.frame(test.dat[, col.ind])

svm.pred <- predict(svm.model, test.dat.final, na.action = na.exclude, probability=TRUE, decision.values = TRUE)
#test.ans <- cbind(attr(svm.pred, "decision.values"), test.dat.final$label, attr(svm.pred, "probabilities")[,1])
#new.ANS <- rbind(new.ANS, test.ans)

attr(svm.pred, "probabilities")
```

#TODO
#make function to analyze prediction probs
```{r eval=F}
predictionProbs <- read_csv("~/Desktop/classify_ap/basalPredictAP_probabilitiesWholeOB.csv") %>% rename("gene" = "WholeAnt")
ggplot(predictionProbs, aes(High, Low)) + geom_point(aes(color=Position))

info <- read_csv("~/Desktop/classify_ap/knowntanwavg.csv")
ppinfo <- left_join(predictionProbs, info, by = "gene") %>% select(gene, High, Low, Position, tenWA:fssnotenWArank) %>% mutate(PosRank = min_rank(tenWA)) %>% arrange(desc(Position))

p2 <- ggplot(predictionProbs, aes(High, Low)) + geom_point(aes(color=Position, alpha = 0.25)) + theme_cowplot() + xlab("Probability of High Basal Activity") + ylab("Probability of Low Basal Activity") + scale_color_manual(values = c("#00BFFF","#FF3030"))

ggExtra::ggMarginal(
  p = p2,
  type = 'histogram',
  margins = 'x',
  size = 1.5,
  colour = 'black',
  groupFill = TRUE
)


```

