---
title: "Merge OBmap inputs"
author: "kanazian"
date: "`r Sys.Date()`"
output: html_document
always_allow_html: true
---
# Goal
Given gencodeV25, STAR aligned, RSEM quantified OBmap readset, output tpm tables of ORs and all intact (OR+TAAR+VMNR) chemosensory genes

#seems that m14s23 is all 0s?  investigate

```{r setup, message=FALSE}
library(tidyverse) #hail hadley
```


# Make a merged file from all inputs
```{r, message=F, warning=F, eval=F}
#prep loading of all files in dir
filedir <- "~/Desktop/obmap/r_analysis/data/starrsem_gen25_inmodel/"
myfiles <- list.files(path = filedir, pattern = "*.genes.results", full.names = T)

#prep sample names
rawnames <- list.files(path = filedir, pattern = "*.genes.results")
splitnames <- str_split(rawnames, ".genes")

keynames <- vector(mode = "character", length = length(rawnames)+1)
for (i in 1:length(rawnames)) {
    keynames[1] <- "ens_raw"
    keynames[i+1] <- splitnames[[i]][1]
}

#setup to make merge files
sample_file <- read_tsv(myfiles[1])

mergetpm <- matrix(ncol = nrow(sample_file), nrow = length(myfiles) + 1)
for (i in 1:length(myfiles)) {
    if (i == 1) {
        genes <- read_tsv(myfiles[i]) %>% select(gene_id) %>% as_vector()
        mergetpm[i,] <- genes
        mergetpm[i+1,] <- read_tsv(myfiles[i]) %>% select(TPM) %>% as_vector()
    } else {
        mergetpm[i+1,] <- read_tsv(myfiles[i]) %>% select(TPM) %>% as_vector()
    }
}

preptpm <- as_tibble(mergetpm) %>% t() %>% as_tibble()
colnames(preptpm) <- keynames

enstrans <- read_csv("~/Desktop/obmap/r_analysis/data/make_TPMmtx/convert_ensID_geneName.csv") %>%
    select(ens_id, gene_name) %>%
    unique()

renamed <- preptpm %>%
    rowwise() %>%
    mutate(ens_id = str_split(ens_raw, "[.]")[[1]][1]) %>%
    ungroup() %>%
    left_join(enstrans, by = "ens_id") %>%
    select(gene_name, everything()) %>%
    unique()

write_csv(renamed, "~/Desktop/obmap/r_analysis/data/make_TPMmtx/samples_over_allgenes_210421.csv")
```


# intermediate outputs
```{r, message=F, warning=F, eval=F}
olfronly <- renamed %>%
    filter(str_detect(gene_name, "Olfr"))
write_csv(olfronly, "~/Desktop/obmap/r_analysis/data/make_TPMmtx/samples_over_allolfrs_210421.csv")

intactonly <- olfronly %>%
    filter(!str_detect(gene_name, "ps"))
write_csv(renamed, "~/Desktop/obmap/r_analysis/data/make_TPMmtx/samples_over_intactolfrs_210421.csv")

taaronly <- renamed %>%
    filter(str_detect(gene_name, "Taar")) %>%
    filter(!str_detect(gene_name, "ps"))
write_csv(renamed, "~/Desktop/obmap/r_analysis/data/make_TPMmtx/samples_over_intacttaars_210421.csv")

vmnonly <- renamed %>%
    filter(str_detect(gene_name, "Vmn")) %>%
    filter(!str_detect(gene_name, "ps"))
write_csv(renamed, "~/Desktop/obmap/r_analysis/data/make_TPMmtx/samples_over_intactvmn_210421.csv")

oldcap1markers <- c("Omp", "Taf1b", "Xist")
oc1mark <- renamed %>%
    filter(gene_name %in% oldcap1markers)

chemo <- bind_rows(intactonly, taaronly, vmnonly, oc1mark)
write_csv(renamed, "~/Desktop/obmap/r_analysis/data/make_TPMmtx/samples_over_intactchemo_210421.csv")
```


#make genes_over_samples files ready for model input
```{r}
old_covars <- read_csv("~/Desktop/obmap/r_analysis/data/3dimOB/model/covarintactchemo2_over_samples_210227.csv") %>%
    select(name:dimrep)

new_file_divide <- old_covars %>% 
    mutate(X1 = str_replace(name, "[.]", "S")) %>%
    select(X1, everything(), -name)

intactflip <- intactonly %>% select(-ens_raw, -ens_id) %>% t()
intactnames <- intactflip[1,]
colnames(intactflip) <- intactnames
intactflip <- intactflip[2:nrow(intactflip),]

chemoflip <- chemo %>% select(-ens_raw, -ens_id) %>% t()
chemonames <- chemoflip[1,]
colnames(chemoflip) <- chemonames
chemoflip <- chemoflip[2:nrow(chemoflip),]

write.csv(intactflip, "~/Desktop/obmap/r_analysis/data/make_TPMmtx/temp_intactflip.csv")
write.csv(chemoflip, "~/Desktop/obmap/r_analysis/data/make_TPMmtx/temp_chemoflip.csv")


intactjoin <- read_csv("~/Desktop/obmap/r_analysis/data/make_TPMmtx/temp_intactflip.csv")%>%
    mutate(X1 = rownames(intactflip)) %>%
    select(X1, everything()) %>%
    rowwise() %>%
    mutate(O290 = Olfr290 + Olfr290_1) %>%
    ungroup() %>%
    select(-Olfr290, -Olfr290_1) %>%
    rename(Olfr290 = O290)

chemojoin <- read_csv("~/Desktop/obmap/r_analysis/data/make_TPMmtx/temp_chemoflip.csv") %>%
    as_tibble() %>%
    mutate(X1 = rownames(chemoflip)) %>%
    select(X1, everything()) %>% 
    rowwise() %>% 
    mutate(O290 = Olfr290 + Olfr290_1) %>%
    ungroup() %>%
    select(-Olfr290, -Olfr290_1) %>%
    rename(Olfr290 = O290)

intactready <- new_file_divide %>%
    left_join(intactjoin, by = "X1") %>%
    rename("name" = "X1")

chemoready <- new_file_divide %>%
    left_join(chemojoin, by = "X1") %>%
    rename("name" = X1)

write.csv(intactready, "~/Desktop/obmap/r_analysis/data/make_TPMmtx/covarintactolfrs_over_samples_210421.csv")
write.csv(chemoready, "~/Desktop/obmap/r_analysis/data/make_TPMmtx/covarintactchemo_over_samples_210421.csv")
```


# examine chemosensory, target, and marker content across samples and dimensions

Initially I determined DTVT weights using total OR tpm but this is potentially problematic due to the fact that each OR gets a different amount of enrichment. Since each slice likely contains different OR transcripts, the total OR TPM could differ based solely on OR composition even if the number of OR transcripts is the same.  

Therefore, it is better to calculate using a single gene, in this case OMP, as a marker for glomerular surface area. One issue with using OMP TPM is that OMP is enriched far more in ORCapSeq2 than ORCapSeq1 while ORs are enriched far more in ORCapSeq1 than ORCapSeq2.
```{r}
allgenes <- read_csv("~/Desktop/obmap/r_analysis/data/make_TPMmtx/covarintactchemo2_over_samples_210421.csv")

olfr_sums <- allgenes %>% select(-Olfr287, -Olfr32) %>% select_if(str_detect(names(.), "^Olfr")) %>% rowSums()
vmn_sums <- allgenes %>% select_if(str_detect(names(.), "^Vmn")) %>% rowSums()
taar_sums <- allgenes %>% select_if(str_detect(names(.), "^Taar")) %>% rowSums()
omp_sums <- allgenes %>% pull(Omp)

#calculate total TPM for gene families in each sample
gene_sum_tb <- allgenes %>%
    select(name:dimrep) %>%
    add_column(sumolfr = olfr_sums,
               sumvmn = vmn_sums,
               sumtaar = taar_sums,
               sumomp = omp_sums)

oc_sum_tb <- gene_sum_tb %>%
    filter(rep < 13)
nc_sum_tb <- gene_sum_tb %>%
    filter(rep >= 13)

# total OR TPM
ggplot(oc_sum_tb %>% filter(!is.na(dimrep))) +
    geom_line(aes(slice, sumolfr, color = as_factor(dimrep))) +
    facet_wrap(~ dim, nrow = 1, ncol = 3)

ggplot(oc_sum_tb %>% filter(dim == "AntPos")) +
    geom_line(aes(slice, sumolfr, color = as_factor(rep))) +
    ggtitle("AntPos replicates - sum of OR tpm")

ggplot(oc_sum_tb %>% filter(dim == "MedLat")) +
    geom_line(aes(slice, sumolfr, color = as_factor(rep))) +
    ggtitle("MedLat replicates - sum of OR tpm")

ggplot(oc_sum_tb %>% filter(dim == "VenDor")) +
    geom_line(aes(slice, sumolfr, color = as_factor(rep))) +
    ggtitle("VenDor replicates - sum of OR tpm")

# OMP TPM
ggplot(oc_sum_tb %>% filter(!is.na(dimrep))) +
    geom_line(aes(slice, sumomp, color = as_factor(dimrep))) +
    facet_wrap(~ dim, nrow = 1, ncol = 3)

ggplot(oc_sum_tb %>% filter(dim == "AntPos")) +
    geom_line(aes(slice, sumomp, color = as_factor(rep))) +
    ggtitle("AntPos replicates - sum of OMP tpm")

ggplot(nc_sum_tb %>% filter(dim == "AntPos")) +
    geom_line(aes(slice, sumomp, color = as_factor(rep))) +
    ggtitle("AntPos replicates - sum of OMP tpm")

ggplot(oc_sum_tb %>% filter(dim == "MedLat")) +
    geom_line(aes(slice, sumomp, color = as_factor(rep))) +
    ggtitle("MedLat replicates - sum of OMP tpm")

ggplot(oc_sum_tb %>% filter(dim == "VenDor")) +
    geom_line(aes(slice, sumomp, color = as_factor(rep))) +
    ggtitle("VenDor replicates - sum of OMP tpm")

#seems like VD14, AP13, AP15 fall outside of their groups
#interestingly ML16 fits in well, looking at my notes I used oldCap instead of newCap for that replicate (14,15,16)
#seems that differences in total OR tpm indicate a difference between capture arrays
#Vmnrs display a similar trend, taars are also similar

ggplot(gene_sum_tb %>% filter(dim == "AntPos")) +
    geom_line(aes(slice, sumtaar, color = as_factor(rep))) +
    ggtitle("X replicates - sum of Taar tpm")

#make normalization factors
#normalization factors for rep are omp of slice/omp of all slices
#normalization factors for dim are 
norm_stats <- gene_sum_tb %>%
    mutate(cap = ifelse(rep < 13, 1, ifelse(rep == 16, 1, 2))) %>%
    group_by(cap, rep) %>% 
    mutate(total_rep_omp = sum(sumomp),
           total_rep_olfr = sum(sumolfr),
           prop_rep_omp = sumomp/total_rep_omp,
           prop_rep_olfr = sumolfr/total_rep_olfr) %>%
    ungroup() %>%
    group_by(cap, dim) %>%
    mutate(total_dim_omp = sum(sumomp),
           total_dim_olfr = sum(sumolfr)) %>%
    ungroup() %>%
    group_by(cap, dim, slice) %>%
    mutate(prop_dim_omp = sum(sumomp)/total_dim_omp,
           prop_dim_olfr = sum(sumolfr)/total_dim_olfr) %>%
    ungroup()

write_csv(norm_stats, "~/Desktop/obmap/r_analysis/data/make_TPMmtx/weights_proptotal_210424.csv")

norm_stats %>% mutate(test = sumolfr*prop_rep_omp) %>% ggplot() + geom_line(aes(slice, test, color = as_factor(rep))) + facet_wrap(~ dim)

norm_stats %>% mutate(test = sumolfr*prop_dim_omp) %>% ggplot() + geom_line(aes(slice, test, color = as_factor(rep))) + facet_wrap(~ dim)


#older DTVT weight stuff
#for each slice along each dim, calculate ratio of slice TPM to both mean and total TPM of dimension
dim_groups <- gene_sum_tb %>% 
    filter(rep < 13) %>% 
    group_by(dim, slice) %>% 
    summarise(meanolfr = mean(sumolfr), .groups = "keep")

APmeanOR <- dim_groups %>% filter(dim == "AntPos") %>% pull(meanolfr) %>% mean()
MLmeanOR <- dim_groups %>% filter(dim == "MedLat") %>% pull(meanolfr) %>% mean()
VDmeanOR <- dim_groups %>% filter(dim == "VenDor") %>% pull(meanolfr) %>% mean()

APsumOR <- dim_groups %>% filter(dim == "AntPos") %>% pull(meanolfr) %>% sum()
MLsumOR <- dim_groups %>% filter(dim == "MedLat") %>% pull(meanolfr) %>% sum()
VDsumOR <- dim_groups %>% filter(dim == "VenDor") %>% pull(meanolfr) %>% sum()

dim_norms <- dim_groups %>% 
    mutate(dim_norms_ratio2mean = case_when(dim == "AntPos" ~ meanolfr/APmeanOR,
                                 dim == "MedLat" ~ meanolfr/MLmeanOR,
                                 dim == "VenDor" ~ meanolfr/VDmeanOR),
           dim_norms_ratio2total = case_when(dim == "AntPos" ~ meanolfr/APsumOR,
                                 dim == "MedLat" ~ meanolfr/MLsumOR,
                                 dim == "VenDor" ~ meanolfr/VDsumOR)) %>%
    select(dim, slice, dim_norms_ratio2mean, dim_norms_ratio2total) %>% #norm_factor will be used as divisor for heatmap section TPM
    ungroup()

rep_norms <- gene_sum_tb %>% 
    select(-sumvmn, -sumtaar) %>% 
    group_by(rep) %>% 
    mutate(meanOR = mean(sumolfr),
           sumOR = sum(sumolfr)) %>% 
    ungroup() %>% 
    mutate(rep_norms_ratio2mean = sumolfr/meanOR, 
           rep_norms_ratio2total = sumolfr/sumOR)

normies <- rep_norms %>%
    left_join(dim_norms, by = c("dim", "slice")) %>%
    select(name:dimrep, contains("_norms"))

norm_trim <- normies %>% 
    select(-rep, -slice, -dim, -dimrep)

#check values
check_norms <- gene_sum_tb %>%
    left_join(norm_trim, by = "name") %>%
    mutate(rep_OR = sumolfr/rep_norms_ratio2mean,
           dim_OR = sumolfr/dim_norms_ratio2mean,
           rep_vmn = sumvmn/rep_norms_ratio2mean,
           dim_vmn = sumvmn/dim_norms_ratio2mean,
           rep_omp = sumomp/rep_norms_ratio2mean,
           dim_omp = sumomp/rep_norms_ratio2mean)

#saveRDS(norm_trim, "~/Desktop/obmap/r_analysis/data/make_TPMmtx/totalORtpm_normalization_justname_210404.RDS")

#rep norms should create straight lines since each slice within a replicate will be perfectly normalized, although OR content within each sample will be variable
ggplot(check_norms) +
    geom_line(aes(slice, rep_OR, color = as_factor(rep))) +
    ggtitle("ORs normalized by replicate")

#dim norms should be relatively more flat, but still chaotic, reminder that dim norms do not include the outlier replicates that used the newer capture array probes
ggplot(check_norms) +
    geom_line(aes(slice, dim_OR, color = as_factor(rep))) +
    ggtitle("ORs normalized by dimension")

#interestingly dim norms seems to work a little better for vmns
ggplot(check_norms %>% filter(dim == "AntPos")) +
    geom_line(aes(slice, rep_vmn, color = as_factor(rep))) +
    ggtitle("Vmnrs normalized by replicate")
ggplot(check_norms %>% filter(dim == "AntPos")) +
    geom_line(aes(slice, dim_vmn, color = as_factor(rep))) +
    ggtitle("Vmnrs normalized by dimension")
```

