---
title: "OB single cell style analysis"
author: "kanazian"
date: "`r Sys.Date()`"
output: github_document
always_allow_html: true
---
# Goal: 
Using positions from single dimensions as clusters, perform scRNAseq cluster analysis to examine gene differential expression between positional clusters and dimensional reduction for positional samples. Following: https://bioconductor.org/packages/release/bioc/vignettes/SC3/inst/doc/SC3.html#de-genes

## Prep tpm data for sc3
```{r setup, warning=F, message=F}
library(raster)
library(SC3)
library(SingleCellExperiment)
library(scater)
library(plotly)
library(cowplot)
library(knitr)
library(patchwork)
library(tidyverse)

#filtered out ORs from heatmap
filterORs <- readRDS("~/Desktop/heatmaps/filterORs_ompdv22no1315_uni2_ect4_below1mingt_zer0.RDS")

allchemo_raw <- read_csv("~/Desktop/obmap/r_analysis/data/make_TPMmtx/covarintactchemo3_over_samples_210429.csv")

allchemo_tpm <- allchemo_raw %>%
  dplyr::select(name:dimrep, contains(c("Olfr", "Taar"))) %>%
  dplyr::select(-filterORs) %>%
  filter(rep != 13) %>%
  filter(rep != 15) %>% 
  filter(name != "M9S23") %>% 
  filter(name != "M14S23")

info <- read_csv("~/Desktop/info_210430.csv") %>%
  mutate(tz_vd2 = case_when(tz_val < 2 ~ "Dorsal",
                            tz_val >= 2 ~ "Ventral",
                            is.na(tz_val) ~ ifelse(tan_zone == "unusual", "Unusual", "LowExp")),
         dv_wavg = 22-vd_wavg,
         dv_rank = min_rank(dv_wavg)) 

set.seed(711)
```


# Prep single cell experiment
```{r}
allchemo_dim <- allchemo_tpm %>% filter(dim == "AntPos")
sample_names <- allchemo_dim$name
covars <- allchemo_dim %>% select(name:dimrep)

allchemo_trim <- allchemo_dim %>% select(-name, -rep, -slice, -dim, -dimrep) %>% t()
colnames(allchemo_trim) <- sample_names
allchemo_trim[1:5,1:5]

allchemo_coldata <- covars %>% select(name, slice)
coldata_ac <- data.frame("cell_type1" = allchemo_coldata$slice)
rownames(coldata_ac) <- allchemo_coldata$name

sce <- SingleCellExperiment(
    assays = list(
        counts = as.matrix(allchemo_trim),
        logcounts = log2(as.matrix(allchemo_trim) + 1)
    ), 
    colData = coldata_ac
)

rowData(sce)$feature_symbol <- rownames(sce)
```


# visualize dimreduction of single dim slices
```{r}
sce_umap <- runUMAP(sce)

umap_rd <- reducedDim(sce_umap) %>% as_tibble()
umap_tibble <- tibble(name = rownames(reducedDim(sce_umap)), 
                      UMAP1 = umap_rd$V1, UMAP2 = umap_rd$V2) %>%
  left_join(covars, by = "name") %>%
  mutate(Section = slice)

umap_allrep <- ggplot(umap_tibble) +
  geom_point(aes(UMAP1, UMAP2, fill = Section),
             size = 5, pch = 21, color = "black") +
  scale_fill_viridis_c() +
  theme_bw() +
  theme_kz() +
  theme(legend.position="none")

umap_byrep <- ggplot(umap_tibble) + 
  geom_vline(xintercept = 0, linetype = "dashed", color = "red") +
  geom_hline(yintercept = 0, linetype = "dashed", color = "red") +
  geom_point(aes(UMAP1, UMAP2, fill = Section),
             size = 3, pch = 21, color = "black") +
  scale_shape_manual(values=c(0, 1, 2)) +
  scale_fill_viridis_c() +
  facet_wrap(~ dimrep) +
  theme_bw() +
  theme_kz() + 
  theme(legend.position = "none") +
  theme(panel.grid.minor = element_blank())

umap_allrep
umap_byrep
```

# umap for all dims
```{r}
set.seed(711)
alldim <- allchemo_tpm %>% filter(rep != 16)
ad_sample_names <- alldim$name
ad_covars <- alldim %>% select(name:dimrep)

ad_trim <- alldim %>% select(-name, -rep, -slice, -dim, -dimrep) %>% t()
colnames(ad_trim) <- ad_sample_names
ad_trim[1:5,1:5]

adcoldata <- ad_covars %>% select(name, slice)
ad_coldata_ac <- data.frame("cell_type1" = adcoldata$slice)
rownames(ad_coldata_ac) <- adcoldata$name

ad_sce <- SingleCellExperiment(
    assays = list(
        counts = as.matrix(ad_trim),
        logcounts = log2(as.matrix(ad_trim) + 1)
    ), 
    colData = ad_coldata_ac)

rowData(ad_sce)$feature_symbol <- rownames(ad_sce)

ad_umap <- runUMAP(ad_sce)
ad_umap_rd <- reducedDim(ad_umap) %>% as_tibble()
ad_tibble <- tibble(name = rownames(reducedDim(ad_umap)), 
                      UMAP_1 = ad_umap_rd$V1, UMAP_2 = ad_umap_rd$V2) %>%
  left_join(ad_covars, by = "name") %>%
  mutate(slice_fct = as_factor(slice),
         dimrep_fct = as_factor(dimrep),
         index = 1:n())

ad_umap_plot <- ggplot(ad_tibble) + 
  geom_point(aes(UMAP_1, UMAP_2, fill = slice_fct),
             size = 6, pch = 21, color = "black") +
  scale_fill_viridis_d() +
  theme_cowplot() +
  ggtitle("UMAP of OR Gene Expression", 
          subtitle = paste0(as.character(dim(ad_trim)[2]), 
                            " alldim sections from 11 mice")) + 
  theme(legend.position = "none") +
  geom_vline(xintercept = 0, linetype = "dashed", color = "red") +
  geom_hline(yintercept = 0, linetype = "dashed", color = "red")

#hiro doesnt like the simple all dim plot colored by slice, make his request
hmt1 <- ad_tibble %>% mutate(facet = "AntPos", 
                             coloredin = ifelse(dimrep == 1, "B", "A"),
                             transp = ifelse(dimrep == 1, 1, 0.9))
hmt2 <- ad_tibble %>% mutate(facet = "MedLat", 
                             coloredin = ifelse(dimrep == 2, "B", "A"),
                             transp = ifelse(dimrep == 2, 1, 0.9))
hmt3 <- ad_tibble %>% mutate(facet = "VenDor", 
                             coloredin = ifelse(dimrep == 3, "B", "A"),
                             transp = ifelse(dimrep == 3, 1, 0.9))
hm12 <- bind_rows(hmt1, hmt2)
hm123 <- bind_rows(hm12, hmt3)

hm_umap_plot <- ggplot(hm123) + 
  geom_point(aes(UMAP_1, UMAP_2, fill = coloredin, alpha = transp),
             size = 6, pch = 21, color = "black") +
  scale_fill_manual(values = c("white", "red")) +
  theme_cowplot() +
  ggtitle("UMAP of OR Gene Expression", 
          subtitle = paste0(as.character(dim(ad_trim)[2]), 
                            " alldim sections from 12 mice")) + 
  theme(legend.position = "none") +
  geom_vline(xintercept = 0, linetype = "dashed", color = "red") +
  geom_hline(yintercept = 0, linetype = "dashed", color = "red") +
  facet_wrap(~ facet)

ad_facets <- ggplot(ad_tibble) + 
  geom_point(aes(UMAP_1, UMAP_2, fill = slice),
             size = 6, pch = 21, color = "black") +
  facet_wrap(~ dim) +
  scale_fill_viridis_c() +
  theme_cowplot() +
  geom_vline(xintercept = 0, linetype = "dashed", color = "red") +
  geom_hline(yintercept = 0, linetype = "dashed", color = "red")

alldimplots <- ad_umap_plot / ad_facets
hmplots <- hm_umap_plot / ad_facets
alldimplots
hmplots
#ggsave(alldimplots, filename = "~/Desktop/obmap/r_analysis/sc_style/alldim_umap.png", device = "png")
```


# find closest sample to each sample in terms of UMAP position
```{r}
ad_xy <- cbind(ad_tibble$UMAP_1, ad_tibble$UMAP_2)
distances <- pointDistance(ad_xy, lonlat = F)
distances[which(distances == 0)] <- NA

#check that no points have identical coordinates
sum(is.na(distances)) == nrow(distances)

bestdist <- vector(mode = "numeric", length = nrow(distances))
for (i in 1:nrow(distances)) {
  bestdist[i] <- which(distances[i,] == min(distances[i,], na.rm=T))
  if (i == 1) {
    best_tibble <- ad_tibble[bestdist[i],]
  } else if (i == nrow(distances)) {
    newline <- ad_tibble[bestdist[i],]
    best_tibble <- bind_rows(best_tibble, newline) %>% rename(bname = name,
                                                    bUMAP_1 = UMAP_1,
                                                    bUMAP_2 = UMAP_2,
                                                    brep = rep,
                                                    bslice = slice,
                                                    bdim = dim,
                                                    bdimrep = dimrep,
                                                    bslice_fct = slice_fct,
                                                    bdimrep_fct = dimrep_fct,
                                                    bindex = index)
  } else {
    newline <- ad_tibble[bestdist[i],]
    best_tibble <- bind_rows(best_tibble, newline)
  }
}

best_friends <- bind_cols(ad_tibble, best_tibble) %>%
  mutate(same_dim = ifelse(dim == bdim, T, F),
         same_rep = ifelse(rep == brep, T, F),
         slice_diff = abs(slice - bslice),
         dimdim = paste(dim, bdim, sep = "_"),
         dist = sqrt((UMAP_1 - bUMAP_1)^2) + (UMAP_2 - bUMAP_2)^2)

ggplot(best_friends) +
  geom_histogram(aes(dist)) +
  geom_vline(xintercept = mean(best_friends$dist), 
             color = "red", linetype = 2) +
  geom_vline(xintercept = median(best_friends$dist), 
             color = "blue", linetype = 2) +
  ggtitle("Histogram of distance between closest points", 
          subtitle = "Median is blue, Mean is red")

ggplot(best_friends) +
  geom_bar(aes(same_dim)) +
  ggtitle("Neighbors are typically from same dimension")

ggplot(best_friends %>% filter(same_dim == T)) +
  geom_bar(aes(same_rep)) +
  ggtitle("Neighbors from same dim are typically not from same replicate")

ggplot(best_friends %>% filter(same_dim == T) %>% filter(same_rep == F)) +
  geom_jitter(aes(slice, bslice)) +
  ggtitle("Neighbors from same dim but diff rep, are typically from similar position")

#simplified from above
ggplot(best_friends %>% 
         filter(same_dim == T) %>% 
         filter(same_rep == F) %>% 
         filter(rep == 4)) +
  geom_jitter(aes(slice, bslice, color = brep)) +
  ggtitle("Neighbors from same dim but diff rep, are typically from similar position")

#what about closest points that are not from same dim?
ggplot(best_friends %>%
         filter(same_dim == F) %>%
         group_by(dimdim) %>% 
         summarise(count = n()) %>% 
         arrange(desc(count)) %>% 
         mutate(dimdimfct = as_factor(dimdim))) +
  geom_bar(aes(dimdimfct, count), stat= "identity") +
  ggtitle("Neighbors from different dims are typically DV/ML hybrids")
```


# make AP UMAP colored by replicate and labeled by position
```{r, eval=F}
ad_tibble <- tibble(name = rownames(reducedDim(ad_umap)), 
                      UMAP_1 = ad_umap_rd$V1, UMAP_2 = ad_umap_rd$V2) %>%
  left_join(ad_covars, by = "name") %>%
  mutate(slice_fct = as_factor(slice),
         dimrep_fct = as_factor(dimrep),
         numlab = ifelse(slice == 1, 
                    1, ifelse(slice == 5,
                         5, ifelse(slice == 10,
                              10, ifelse(slice == 15,
                                    15, ifelse(slice == 20, 
                                          20, ifelse(slice == 23, 
                                                     23, NA)))))))

ad_umap_plot <- ggplot(ad_tibble) + 
  geom_point(aes(UMAP_1, UMAP_2, fill = dimrep_fct),
             size = 6, pch = 21, color = "black") +
  geom_label_repel(aes(UMAP_1, UMAP_2, label = numlab)) + 
  scale_fill_viridis_d() +
  theme_cowplot() +
  ggtitle("UMAP of OR Gene Expression", 
          subtitle = paste0(as.character(dim(ad_trim)[2]), 
                            " AP sections from 6 mice")) + 
  theme(legend.position = "none") +
  geom_vline(xintercept = 0, linetype = "dashed", color = "red") +
  geom_hline(yintercept = 0, linetype = "dashed", color = "red")  +
  facet_wrap(~ dimrep)
```

## previously run UMAP for AP dimension
Very ordered set of points reflects the distinct nature of OR positions along the AP axis with the 2 glomeruli positioned in approximately located sections.
```{r, out.width = "50%", out.height = "50%"}
include_graphics("AP_umap.png")
```


## previously run UMAP for ML dimension
Lack of order is expected as the 2 glomeruli per OR can be located in very far apart or very close sections based on distance to mirror symmetry line that separates the ML dimension.
```{r, out.width = "50%", out.height = "50%"}
include_graphics("ML_umap.png")
```


## previously run UMAP for VD dimension
Order is expected as both glomeruli should be located in the same VD section since OB position is related to OE position
```{r, out.width = "50%", out.height = "50%"}
include_graphics("VD_umap.png")
```


## previously run UMAP for all dimension
```{r, include=F, out.width = "50%", out.height = "50%"}
include_graphics("alldim_umap.png")
```


# run sc3
Interactive is very good, but keeping plot code for output
```{r}
sce_sc3 <- sc3(sce, ks = 3:10, biology = TRUE)
#sc3_interactive(sce_sc3)
sc3_export_results_xls(sce_sc3, filename = "~/Desktop/obmap/r_analysis/sc_style/output/ap_sc3.xls")

plotPCA(
    sce_sc3, 
    colour_by = "sc3_5_clusters", 
    size_by = "sc3_5_log2_outlier_score")

sc3_plot_consensus(
    sce_sc3, k = 5, 
    show_pdata = c(
        "cell_type1", 
        "log10_total_features",
        "sc3_5_clusters", 
        "sc3_5_log2_outlier_score"))

sc3_plot_silhouette(sce_sc3, k = 5)

sc3_plot_expression(sce_sc3, k = 5)

sc3_plot_cluster_stability(sce_sc3, k = 5)

sc3_plot_de_genes(
    sce_sc3, k = 5, 
    show_pdata = c(
        "cell_type1", 
        "log10_total_features",
        "sc3_5_clusters", 
        "sc3_5_log2_outlier_score"))

sc3_plot_markers(
    sce_sc3, k = 5, 
    show_pdata = c(
        "cell_type1", 
        "log10_total_features",
        "sc3_3_clusters", 
        "sc3_3_log2_outlier_score"))
```


# Prep single cell experiment for ORs as samples
```{r}
or_dim <- allchemo_tpm %>% select(starts_with("Olfr"))
or_names <- colnames(or_dim)

or_trim <- or_dim %>% as.matrix()
colnames(or_trim) <- or_names
or_trim[1:5,1:5]

coldata_or <- data.frame("cell_type1" = c(1:length(or_names)))
rownames(coldata_or) <- or_names

or_sce <- SingleCellExperiment(
    assays = list(
        counts = as.matrix(or_trim),
        logcounts = log2(as.matrix(or_trim) + 1)
    ), 
    colData = coldata_or
)

rowData(or_sce)$feature_symbol <- rownames(or_sce)

#UMAP
or_umap <- runUMAP(or_sce)
plotUMAP(or_umap, colour_by = "cell_type1")


lancet <- read_csv("~/Downloads/lancet_mouseORnames.csv")

or_rd <- reducedDim(or_umap) %>% as_tibble()
or_umap_tibble <- tibble(Olfrname = rownames(reducedDim(or_umap)), 
                      UMAP_1 = or_rd$V1, UMAP_2 = or_rd$V2) %>%
  left_join(lancet, by = "Olfrname") %>% 
  rename(olfrname = Olfrname) %>%
  left_join(info, by = "olfrname") %>%
  select(olfrname:class, fisurface, tan_zone) %>%
  mutate(symfam = as.numeric(str_extract(Symbol, "[0-9]+")),
         orsort = 1:n()) %>%
  arrange(class) %>%
  mutate(csort = 1:n()) %>%
  arrange(symfam) %>%
  mutate(sfsort = 1:n()) %>%
  arrange(orsort) %>%
  filter(!is.na(class)) %>%
  filter(!is.na(symfam))

ggplot(or_umap_tibble) + 
  geom_point(aes(UMAP_1, UMAP_2, fill = as_factor(symfam)),
             size = 2, pch = 21, color = "black") +
  scale_fill_viridis_d() +
  theme_cowplot() +
  ggtitle("ORs by family (color)", 
          subtitle = "Is that a fish?") + 
  geom_vline(xintercept = 0, linetype = "dashed", color = "red") +
  geom_hline(yintercept = 0, linetype = "dashed", color = "red") +
  theme(legend.position = "none")
```

