---
title: "wachR2_gen25"
author: "kanazian"
date: "`r Sys.Date()`"
output: github_document
always_allow_html: true
---

Goal: Re-examine the differential expression analysis of Functional Imaging OB Surface vs non-FIsurface samples from Wachowiak lab (second set).
```{r setup, message=FALSE}
library(edgeR)
library(DESeq)
library(merTools)
library(plotly)
library(tidyverse)
library(ggrepel)
library(cowplot)

#Set working directory here
#setwd("~/Desktop/obmap/r_analysis/func_surface/")

#Differential Expression function
DiffExp <- function (targets, countsTable) {
  Treat <- factor(targets$Treatment)
  Subject <- factor(targets$Subject)
  design <- model.matrix(~Subject+Treat)
  cds <- newCountDataSet(countsTable,Treat)
  cds <- estimateSizeFactors(cds)
  cds <- estimateDispersions(cds)
  d <- nbinomTest(cds,"0","1")
  e.litter <- DGEList(counts=countsTable)
  e.litter <- estimateGLMCommonDisp(e.litter,design)
  e.litter <- estimateGLMTrendedDisp(e.litter,design)
  e.litter <- estimateGLMTagwiseDisp(e.litter,design)
  fit <- glmFit(e.litter, design)
  lrt <- glmLRT(fit)
  diff <- topTags(lrt,n=dim(lrt)[1])$table
  result <- merge(merge(diff,countsTable,by=0,sort=F), 
                  d, by.x="Row.names", by.y="id", sort=F)
  colnames(result)[1] <- "id"
  return(result)
} #end DiffExp

info <- read_csv("~/Desktop/info_210430.csv") %>%
  rename("gene_name" = "olfrname")
```


## Load files and run DiffE
```{r}
#group1 - nonFI surface/ventral OB
g1a <- read.table("~/Desktop/obmap/r_analysis/data/func_surface/sr_gen25_out/1LV_wR2.genes.results", header=T)
g1b <- read.table("~/Desktop/obmap/r_analysis/data/func_surface/sr_gen25_out/1RV_wR2.genes.results", header=T)
g1c <- read.table("~/Desktop/obmap/r_analysis/data/func_surface/sr_gen25_out/2LV_wR2.genes.results", header=T)
g1d <- read.table("~/Desktop/obmap/r_analysis/data/func_surface/sr_gen25_out/2RV_wR2.genes.results", header=T)
g1e <- read.table("~/Desktop/obmap/r_analysis/data/func_surface/sr_gen25_out/3LV_wR2.genes.results", header=T)
g1f <- read.table("~/Desktop/obmap/r_analysis/data/func_surface/sr_gen25_out/3RV_wR2.genes.results", header=T)
g1g <- read.table("~/Desktop/obmap/r_analysis/data/func_surface/sr_gen25_out/4LV_wR2.genes.results", header=T)
g1h <- read.table("~/Desktop/obmap/r_analysis/data/func_surface/sr_gen25_out/4RV_wR2.genes.results", header=T)

#group2 - FI surface/dorsal OB
g2a <- read.table("~/Desktop/obmap/r_analysis/data/func_surface/sr_gen25_out/1LD_wR2.genes.results", header=T)
g2b <- read.table("~/Desktop/obmap/r_analysis/data/func_surface/sr_gen25_out/1RD_wR2.genes.results", header=T)
g2c <- read.table("~/Desktop/obmap/r_analysis/data/func_surface/sr_gen25_out/2LD_wR2.genes.results", header=T)
g2d <- read.table("~/Desktop/obmap/r_analysis/data/func_surface/sr_gen25_out/2RD_wR2.genes.results", header=T)
g2e <- read.table("~/Desktop/obmap/r_analysis/data/func_surface/sr_gen25_out/3LD_wR2.genes.results", header=T)
g2f <- read.table("~/Desktop/obmap/r_analysis/data/func_surface/sr_gen25_out/3RD_wR2.genes.results", header=T)
g2g <- read.table("~/Desktop/obmap/r_analysis/data/func_surface/sr_gen25_out/4LD_wR2.genes.results", header=T)
g2h <- read.table("~/Desktop/obmap/r_analysis/data/func_surface/sr_gen25_out/4RD_wR2.genes.results", header=T)

#make table, takes a couple mins
countsTable <- round(data.frame(one1=g1a$expected_count, 
                                one2=g1b$expected_count, 
                               one3=g1c$expected_count, 
                               one4=g1d$expected_count, 
                               one5=g1e$expected_count, 
                               one6=g1f$expected_count, 
                               one7=g1g$expected_count, 
                               one8=g1h$expected_count, 
                               two1=g2a$expected_count, 
                               two2=g2b$expected_count, 
                               two3=g2c$expected_count,  
                               two4=g2d$expected_count, 
                               two5=g2e$expected_count, 
                               two6=g2f$expected_count, 
                               two7=g2g$expected_count, 
                               two8=g2h$expected_count, 
                               row.names=g1a$gene_id))

#Use this line if you want to run DiffE using only Olfr genes
#countsTable2 <- countsTable[which(str_detect(rownames(countsTable), "Olfr")),]

#make the labels, needs to match countsTable
diffE_labels <- data.frame(FileName=c("one1","one2","one3","one4","one5","one6","one7","one8", 
                                      "two1","two2","two3","two4","two5","two6","two7","two8"),
                           Subject=c("A","B","C","D","E","F","G","H",
                                     "A","B","C","D","E","F","G","H"),
                           Treatment=c(0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1))

#run differential expression with all genes, check dimension of dataframe and view head
diffE_result <- DiffExp(diffE_labels, countsTable) %>% rename("gene_id" = "id")
#dim(diffE_result)
diffE_result[1:5,1:19]
#id = gene name
#left side is edgeR output
#logFC = log fold change (Between group1 and group2)
#logCPM = log Counts Per Million
#Pvalue = raw significance
#FDR = multiple comparison corrected significance - use this, typically with 0.05 to start
#one1:two3 = counts per sample
#right side is DEseq output
#baseMean/A/B = Mean of counts
#padj = DEseq MCcorrected significance
```


## Prep results and plot all genes in volcano
```{r message=FALSE}
#tibbles are nicer dataframes
der <-as_tibble(diffE_result)

###I hate dealing with attributes so will reload the file as tibble and then left_join names
write_csv(der, "~/Downloads/temp_results.csv")
der <- read_csv("~/Downloads/temp_results.csv") %>%
  rowwise() %>%
  mutate(trim_id = str_split(gene_id, "[.]")[[1]][1]) %>%
  select(-gene_id) %>%
  select(trim_id, everything()) %>%
  rename(gene_id = trim_id) %>%
  ungroup()

#change ensembl_geneIDs to gene names
#since this conversion file also have ensembl transcript IDs, remove duplicates first
ens_names <- read_csv("~/Desktop/obmap/r_analysis/inputs/convert_ensID_geneName.csv") %>% 
  select(ens_id, gene_name) %>% 
  rename("gene_id" = "ens_id") %>%
  unique()

der_named <- left_join(der, ens_names, by = "gene_id") %>% 
  select(gene_name, logFC:padj) %>% 
  rowwise() %>%
  mutate(isOlfr = ifelse(str_detect(gene_name, "Olfr"), T, F),
         isTaar = ifelse(str_detect(gene_name, "Taar"), T, F),
         isVmn = ifelse(str_detect(gene_name, "Vmn"), T, F),
         isPS = ifelse(str_detect(gene_name, "-ps"), T, F),
         isChemo = ifelse(sum(isOlfr, isTaar, isVmn) > 0, T, F),
         mkate2OR = ifelse(gene_name == "Olfr881", "Olfr881", 
                              ifelse(gene_name == "Olfr1377", 
                                     "Olfr1377", "Olfr"))) %>%
  ungroup()

write_csv(der_named, "~/Desktop/obmap/r_analysis/func_surface/allgenes_results.csv")

#quick volcano plot using edgeR
ggplot(der_named) + 
  geom_point(aes(logFC,-log10(FDR), alpha = 0.25, color = isOlfr)) + 
  geom_vline(xintercept = 0)  + 
  geom_hline(yintercept = -log10(0.05), linetype = 2) + 
  scale_color_manual(values = c("#000000","#FF0000"))

##interactively to ask where do ORs fall among all genes
plot_ly(der_named, x = ~logFC, y = ~-log10(FDR), 
        color = ~isOlfr, 
        text = ~paste("ID: ", gene_name, 
                      "<br>logFC: ", logFC,                                                             "<br>FDR", FDR))
```


## plot only chemoreceptors and calculate new FDR for subsets of all genes
```{r}
#itnact chemoreceptors
only_chemo <- der_named %>% 
  filter(isChemo == T) %>%
  filter(isPS == F)

only_or <- only_chemo %>% 
  filter(isOlfr == T)
only_taar <- only_chemo %>% 
  filter(isTaar == T)
only_vmn <- only_chemo %>% 
  filter(isVmn == T)

#recalculate FDR
only_or$FDRor <- p.adjust(only_or$PValue, method = "fdr")
only_taar$FDRtaar <- p.adjust(only_taar$PValue, method = "fdr")
only_vmn$FDRvmn <- p.adjust(only_vmn$PValue, method = "fdr")

makers <- only_or %>% mutate(fisurface = ifelse(logFC > 0, ifelse(FDRor < 0.05, T, F), F))
fisigor <- only_or$gene_name[which(only_or$logFC > 0 & only_or$FDRor < 0.05)]
only_taar$gene_name[which(only_taar$logFC > 0 & only_taar$FDRtaar < 0.05)]
only_vmn$gene_name[which(only_vmn$logFC > 0 & only_vmn$FDRvmn < 0.05)]

write_csv(only_or, 
          "~/Desktop/obmap/r_analysis/func_surface/olfr_fdradj_201123.csv")
write_csv(only_taar, 
          "~/Desktop/obmap/r_analysis/func_surface/taar_fdradj_201123.csv")
write_csv(only_vmn, 
          "~/Desktop/obmap/r_analysis/func_surface/vmn_fdradj_201123.csv")

only_or %>% 
  filter(isOlfr == T) %>%
  left_join(info, by = "gene_name") %>%
  mutate(class_OR = ifelse(class == 1, "Class1", "Class2")) %>%
  arrange(desc(class_OR)) %>%
  ggplot(aes(logFC, -log10(FDRor), color = class_OR)) +
  geom_point() + 
  geom_vline(xintercept = 0)  + 
  geom_hline(yintercept = -log10(0.05), linetype = 2) +
  ggtitle("Olfrs from All Genes DE", subtitle = "FDR re-calculated") +
  xlab("Not FI surface  <<<   logFC   >>>     FI surface")

only_taar %>% 
  filter(isTaar == T) %>%
  ggplot(aes(logFC, -log10(FDRtaar), alpha = 0.25)) + 
  geom_point() + 
  geom_vline(xintercept = 0) + 
  geom_hline(yintercept = -log10(0.05)) +
  ggtitle("Taars from All Genes DE", subtitle = "FDR re-calculated") +
  xlab("Not FI surface  <<<   logFC   >>>     FI surface")

only_vmn %>% 
  filter(isVmn == T) %>%
  ggplot(aes(logFC, -log10(FDRvmn), alpha = 0.25)) + 
  geom_point() + 
  geom_vline(xintercept = 0) + 
  geom_hline(yintercept = -log10(0.05)) +
  ggtitle("Vmnrs from All Genes DE", subtitle = "FDR re-calculated") +
  xlab("Not FI surface  <<<   logFC   >>>     FI surface")

plot_ly(only_chemo, x = ~logFC, y = ~-log10(FDR), 
        color = ~isTaar,
        text = ~paste("ID: ", gene_name,
                      "<br>logFC: ", logFC,
                      "<br>FDR", FDR))

info <- read_csv("~/Desktop/obmap/r_analysis/3dimOB/input/knowntanwavgFI.csv") %>%
  rename("gene_name" = "gene") %>%
  select(gene_name:RTP)

feats <- only_or %>% 
  select(gene_name:FDR, FDRor, isOlfr:mkate2OR) %>% 
  left_join(info, by = "gene_name") %>%
  filter(isPS == F)

# how many from each class
feats %>% 
  filter(FDRor < 0.05) %>%
  filter(logFC > 0) %>%
  group_by(class) %>%
  summarise(count = n())

# how many from each OE region
feats %>% 
  filter(FDRor < 0.05) %>%
  filter(logFC > 0) %>%
  group_by(oe_region) %>%
  summarise(count = n())

# how many TAARs enriched in FI surface
only_taar %>% 
  filter(FDRtaar < 0.05) %>%
  filter(logFC > 0) %>% 
  select(gene_name, PValue, FDR, FDRtaar)
```


# Subset count table to just ORs then run DE
```{r, eval=F}
gene_list <- as.character(g1a$gene_id) %>% 
  as_tibble() %>% 
  rename("gene_id" = "value") %>% 
  rowwise() %>% 
  mutate(trim_id = str_split(gene_id, "[.]")[[1]][1]) %>% 
  rename("dot_id" = "gene_id", "gene_id" = "trim_id") %>% 
  left_join(ens_names, by = "gene_id") %>% 
  mutate(isOlfr = str_detect(gene_name, "Olfr")) %>% 
  filter(isOlfr == T)

orct <- countsTable[which(rownames(countsTable) %in% gene_list$dot_id),]

#run differential expression, check dimension of dataframe and view head
olfr_result <- DiffExp(diffE_labels, orct) %>% 
  as_tibble() %>% 
  rename("gene_id" = "id") %>% 
  mutate(isSignif = ifelse(FDR < 0.05, "FDR<0.05", "Notsig")) %>%
  rowwise() %>% 
  mutate(trim_id = str_split(gene_id, "[.]")[[1]][1]) %>% 
  rename("dot_id" = "gene_id", "gene_id" = "trim_id") %>% 
  left_join(ens_names, by = "gene_id") %>%
  select(gene_name, logFC:isSignif) %>%
  mutate(FI_sig = ifelse(FDR < 0.05 & logFC > 0, T, F),
         RM_sig = ifelse(FDR < 0.05 & logFC < 0, T, F),
         mkate2OR = ifelse(gene_name == "Olfr881", "Olfr881", 
                              ifelse(gene_name == "Olfr1377", "Olfr1377", 
                                     "Olfr")))

olfr_result %>% filter(FI_sig == T)

ggplot(olfr_result) + 
  geom_point(aes(logFC,-log10(FDR), alpha = 0.25, color = mkate2OR)) + 
  geom_vline(xintercept = 0) + 
  geom_hline(yintercept = -log10(0.05))

write_csv(olfr_result, "~/Desktop/obmap/r_analysis/func_surface/orOnlyDiffE_info.csv")
```


# R4.0 has no DEseq so cant rerun DiffE for now.  Loading dataframes and remaking plots
```{r}
allgenes <- read_csv("~/Desktop/obmap/r_analysis/data/func_surface/allgenes_results_201123.csv")
oronly <- read_csv("~/Desktop/obmap/r_analysis/data/func_surface/orOnlyDiffE_info_201123.csv") %>%
  mutate(newFDR = FDR*0.397) %>%
  dplyr::select(-FDR) %>%
  rename(FDR = newFDR) #adj already applied

taaradj <- read_csv("~/Desktop/obmap/r_analysis/data/func_surface/taar_fdradj_201123.csv")
vmnadj <- read_csv("~/Desktop/obmap/r_analysis/data/func_surface/vmn_fdradj_201123.csv")
olfradj <- read_csv("~/Desktop/obmap/r_analysis/data/func_surface/olfr_fdradj_201123.csv")

methoxy <- read_csv("~/Desktop/obmap/r_analysis/data/func_surface/pS6_DE_1p_4methylAC.csv") %>%
  dplyr::select(gene_symbol, logFC, PValue:ST3, ensembl_gene_id) %>%
  rename("gene_name" = "gene_symbol") %>%
  mutate(mkate2OR = case_when(str_detect(gene_name, "Olfr1377") ~ "Olfr1377",
                              str_detect(gene_name, "Olfr881") ~ "Olfr881",
                              str_detect(gene_name, "Olfr") ~ "Olfr"),
         newfdr = p.adjust(PValue, method = "fdr"))


olfr_from_oronly_green1377_blue881 <- ggplot(oronly) + 
  geom_point(data = oronly %>% filter(mkate2OR == "Olfr"), 
             aes(logFC,-log10(FDR), size = 2, alpha = 0.6), color = "#69747B") +
  geom_point(data = oronly %>% filter(mkate2OR == "Olfr1377"), 
             aes(logFC,-log10(FDR), size = 2), color = "#73B404") + 
  geom_point(data = oronly %>% filter(mkate2OR == "Olfr881"), 
             aes(logFC,-log10(FDR), size = 2), color = "#4BC2D1") +
  geom_vline(xintercept = 0) + 
  geom_hline(yintercept = -log10(0.05), linetype = 2, color = "black") +
  labs(x = "log2(Fold Change)", y = "-log10(FDR)") +
  theme_bw() +
  theme_kz() +
  xlim(-13, 13) +
  theme(legend.position = "none")

olfr_from_oronly_redclass1 <- ggplot(oronlyinfo) + 
  geom_point(data = oronlyinfo %>% filter(class == 2), 
             aes(logFC,-log10(FDR), size = 2), color = "#69747B", alpha = 0.6) + 
  geom_point(data = oronlyinfo %>% filter(class == 1),
             aes(logFC,-log10(FDR), size = 2), color = "#ff0000", alpha = 0.5) + 
  geom_vline(xintercept = 0) + 
  geom_hline(yintercept = -log10(0.05), linetype = 2, color = "black") +
  labs(x = "log2(Fold Change)", y = "-log10(FDR)") +
  theme_bw() +
  theme_kz() +
  xlim(-13, 13) +
  theme(legend.position = "none")

taar_from_allgenes <- ggplot(taaradj) + 
  geom_point(aes(logFC,-log10(FDR), size = 2, alpha = 0.6), color = "#69747B") +
  geom_vline(xintercept = 0) + 
  geom_hline(yintercept = -log10(0.05), linetype = 2, color = "black") +
  labs(x = "log2(Fold Change)", y = "-log10(FDR)") +
  theme_bw() +
  theme_kz() +
  xlim(-12, 12) +
  theme(legend.position = "none")

vmn_from_allgenes <- ggplot(vmnadj) + 
  geom_point(aes(logFC,-log10(FDR), size = 2, alpha = 0.6), color = "#69747B") +
  geom_vline(xintercept = 0) + 
  geom_hline(yintercept = -log10(0.05), linetype = 2, color = "black") +
  labs(x = "log2(Fold Change)", y = "-log10(FDR)") +
  theme_bw() +
  theme_kz() +
  xlim(-12, 12) +
  theme(legend.position = "none")

allgenes_orhighlight <- ggplot(allgenes) + 
  geom_point(aes(logFC,-log10(FDR), color = isOlfr, size = 2), alpha = 0.6) +
  scale_color_manual(values = c("#69747B", "#ff0000")) +
  geom_vline(xintercept = 0) + 
  geom_hline(yintercept = -log10(0.05), linetype = 2, color = "black") +
  labs(x = "log2(Fold Change)", y = "-log10(FDR)") +
  theme_bw() +
  theme_kz() +
  xlim(-14, 14) +
  theme(legend.position = "none")

allgenes_orhighlight_legend <- ggplot(allgenes) + 
  geom_point(aes(logFC,-log10(FDR), color = isOlfr, size = 2), alpha = 0.6) +
  scale_color_manual(values = c("#69747B", "#ff0000")) +
  geom_vline(xintercept = 0) + 
  geom_hline(yintercept = -log10(0.05), linetype = 2, color = "black") +
  labs(x = "log2(Fold Change)", y = "-log10(FDR)") +
  theme_bw() +
  theme_kz() +
  xlim(-14, 14)

methoxy_green1377_blue881 <- ggplot() + 
  geom_point(data = methoxy %>% filter(mkate2OR == "Olfr"), 
             aes(logFC,-log10(FDR), size = 2, alpha = 0.6), color = "#69747B") +
  geom_point(data = methoxy %>% filter(mkate2OR == "Olfr1377"), 
             aes(logFC,-log10(FDR), size = 2), color = "#73B404") + 
  geom_point(data = methoxy %>% filter(mkate2OR == "Olfr881"), 
             aes(logFC,-log10(FDR), size = 2), color = "#4BC2D1") +
  geom_vline(xintercept = 0) + 
  geom_hline(yintercept = -log10(0.05), linetype = 2, color = "black") +
  labs(x = "log2(Fold Change)", y = "-log10(FDR)") +
  theme_bw() +
  theme_kz() +
  xlim(-8, 8) +
  theme(legend.position = "none")

oronly %>% filter(FDR < 0.05) %>% filter(logFC > 0) %>% count()

SaveRectPlots(olfr_from_oronly_green1377_blue881, olfr_from_oronly_redclass1, taar_from_allgenes, vmn_from_allgenes, allgenes_orhighlight_legend, allgenes_orhighlight, methoxy_green1377_blue881, path = "~/Desktop/func_surface")




```

