---
title: "OBmap single cell style analysis"
author: "kanazian"
date: "Sep 24, 2020"
output: 
  html_document: 
    keep_md: true
---
# Goal: 
Perform scRNAseq cluster DiffE style analysis on replicates of single dimension OBmap data to examine gene differential expression between positional clusters and dimensional reduction for positional samples. Following: https://bioconductor.org/packages/release/bioc/vignettes/SC3/inst/doc/SC3.html#de-genes

## Prep tpm data for sc3
```{r setup, include=FALSE}
library(SC3)
library(SingleCellExperiment)
library(scater)
library(plotly)
library(tidyverse)
library(cowplot)

allchemo_tpm <- read_csv("~/Desktop/obmap/r_analysis/3dimOB/input/covarintactchemo_over_samples_200923.csv") %>% select(-X1)

set.seed(711)
```


#Prep single cell experiment
```{r}
allchemo_dim <- allchemo_tpm %>% filter(dim == "AntPos")
sample_names <- allchemo_dim$name
covars <- allchemo_dim %>% select(name:dimrep)

allchemo_trim <- allchemo_dim %>% select(-name, -rep, -slice, -dim, -dimrep) %>% t()
colnames(allchemo_trim) <- sample_names
allchemo_trim[1:5,1:5]

allchemo_coldata <- covars %>% select(name, slice)
coldata_ac <- data.frame("cell_type1" = allchemo_coldata$slice)
rownames(coldata_ac) <- allchemo_coldata$name

sce <- SingleCellExperiment(
    assays = list(
        counts = as.matrix(allchemo_trim),
        logcounts = log2(as.matrix(allchemo_trim) + 1)
    ), 
    colData = coldata_ac
)

rowData(sce)$feature_symbol <- rownames(sce)
```


#visualize dimreduction of the ~24 AP slices
```{r}
sce_pca <- runPCA(sce)
plotPCA(sce_pca, colour_by = "cell_type1")

sce_tsne <- runTSNE(sce)
plotTSNE(sce_tsne, colour_by = "cell_type1")

sce_umap <- runUMAP(sce)
plotUMAP(sce_umap, colour_by = "cell_type1")

umap_rd <- reducedDim(sce_umap) %>% as_tibble()
umap_tibble <- tibble(name = rownames(reducedDim(sce_umap)), 
                      UMAP_1 = umap_rd$V1, UMAP_2 = umap_rd$V2) %>%
  left_join(covars, by = "name") %>%
  mutate(slice_fct = as_factor(slice))

umapplot <- ggplot(umap_tibble) + 
  geom_point(aes(UMAP_1, UMAP_2, fill = slice),
             size = 6, pch = 21, color = "black") +
  scale_fill_viridis_c() +
  theme_cowplot() +
  ggtitle("UMAP of OR Gene Expression", 
          subtitle = paste0(as.character(dim(allchemo_trim)[2]), 
                            " Anterior-Posterior sections from 6 mice"))
ggsave(umapplot, filename = "~/Desktop/obmap/r_analysis/sc_style/output/ap_umap.png", device = "png")
umapplot

sce_umapcomp <- runUMAP(sce, ncomponents = 5)
plotUMAP(sce_umapcomp, ncomponents = 5, colour_by = "cell_type1")
```


# run sc3
Interactive is very good, but keeping plot code for output
```{r}
sce_sc3 <- sc3(sce, ks = 3:10, biology = TRUE)
#sc3_interactive(sce_sc3)
sc3_export_results_xls(sce_sc3, filename = "~/Desktop/obmap/r_analysis/sc_style/output/ap_sc3.xls")

plotPCA(
    sce_sc3, 
    colour_by = "sc3_5_clusters", 
    size_by = "sc3_5_log2_outlier_score")

sc3_plot_consensus(
    sce_sc3, k = 5, 
    show_pdata = c(
        "cell_type1", 
        "log10_total_features",
        "sc3_5_clusters", 
        "sc3_5_log2_outlier_score"))

sc3_plot_silhouette(sce_sc3, k = 5)

sc3_plot_expression(sce_sc3, k = 5)

sc3_plot_cluster_stability(sce_sc3, k = 5)

sc3_plot_de_genes(
    sce_sc3, k = 5, 
    show_pdata = c(
        "cell_type1", 
        "log10_total_features",
        "sc3_5_clusters", 
        "sc3_5_log2_outlier_score"))

sc3_plot_markers(
    sce_sc3, k = 5, 
    show_pdata = c(
        "cell_type1", 
        "log10_total_features",
        "sc3_3_clusters", 
        "sc3_3_log2_outlier_score"))
```
