---
title: "seqdepth_cdsutr_alignments"
author: "kanazian"
date: "9/21/2020"
output: html_document
---

```{r setup}
library(tidyverse)
```


# Testing sequence depth of capture with 10% sampling of old reads
using 3 samples from 3 different mice, compare tpmcutoff group vs tpmcutoff group from 10% sample and below cutoff group vs below cutoff group from 10% sample
```{r}
#M8 = MedLat
m8s15_10p <- read_tsv("~/Desktop/obmap/heatmaps_starrsem/M8-15_10p.genes.results") %>% select(gene_id, TPM)
#M9 = VenDor
m9s7_10p <- read_tsv("~/Desktop/obmap/heatmaps_starrsem/M9-07_10p.genes.results") %>% select(gene_id, TPM)
#M10 = AntPos
m10s18_10p <- read_tsv("~/Desktop/obmap/heatmaps_starrsem/M10-18_10p.genes.results") %>% select(gene_id, TPM)

FilterOlfr <- function(df) {
  olfrs <- df[which(str_detect(df$gene_id, "Olfr")),]
  olfrs_nops <- olfrs[which(str_detect(olfrs$gene_id, "-ps", negate = T)),] %>% arrange(gene_id)
  olfrs_nopsnoect <- olfrs_nops %>% 
    filter(gene_id != "Olfr32") %>% 
    filter(gene_id != "Olfr287")
  return(olfrs_nopsnoect)
}

m8s15_olfrs <- FilterOlfr(m8s15_10p) %>% rename("m8s15_10" = "TPM")
m9s7_olfrs <- FilterOlfr(m9s7_10p) %>% rename("m9s7_10" = "TPM")
m10s18_olfrs <- FilterOlfr(m10s18_10p) %>% rename("m10s18_10" = "TPM")

#setup 100% reads aligned with older star reference, some ORs removed
#10p shuffled done with newest reference, some ORs added
m8s15_100 <- allmice_tpm %>% filter(rep == 8 & slice == 15) %>% 
  select(-name, -rep, -slice, -dim, -dimrep) %>% t()
colnames(m8s15_100) <- "tpm"
m9s7_100 <- allmice_tpm %>% filter(rep == 9 & slice == 7) %>% 
  select(-name, -rep, -slice, -dim, -dimrep) %>% t()
colnames(m9s7_100) <- "tpm"
m10s18_100 <- allmice_tpm %>% filter(rep == 10 & slice == 18) %>% 
  select(-name, -rep, -slice, -dim, -dimrep) %>% t()
colnames(m10s18_100) <- "tpm"
gene_id <- rownames(m8s15_100)

m8s15_100_tb <- data.frame(gene_id, m8s15_100) %>% as_tibble()
m9s7_100_tb <- data.frame(gene_id, m9s7_100) %>% as_tibble()
m10s18_100_tb <- data.frame(gene_id, m10s18_100) %>% as_tibble()

gene_id <- rownames(m8s15_100)
fullreads <- tibble(gene_id, m8s15_100, m9s7_100, m10s18_100)

ap_upcut <- colnames(ap_tpm %>% select(-name, -rep, -slice, -dim, -dimrep))
ap_downcut <- colnames(ap_tpm[,which(or_mean < 24)])
ml_upcut <- colnames(ml_tpm_cut %>% select(-name, -rep, -slice, -dim, -dimrep))
ml_downcut <- colnames(ml_tpm[,which(ml_or_mean < 16)])
vd_upcut <- colnames(vd_tpm_cut %>% select(-name, -rep, -slice, -dim, -dimrep))
vd_downcut <- colnames(vd_tpm[,which(vd_or_mean < 20)])

above_cut <- "up"
below_cut <- "down"
ap_uplab <- tibble(ap_upcut, above_cut) %>% rename("gene_id" = "ap_upcut", "cutoff" = "above_cut")
ap_downlab <- tibble(ap_downcut, below_cut) %>% rename("gene_id" = "ap_downcut", "cutoff" = "below_cut")
ml_uplab <- tibble(ml_upcut, above_cut) %>% rename("gene_id" = "ml_upcut", "cutoff" = "above_cut")
ml_downlab <- tibble(ml_downcut, below_cut) %>% rename("gene_id" = "ml_downcut", "cutoff" = "below_cut")
vd_uplab <- tibble(vd_upcut, above_cut) %>% rename("gene_id" = "vd_upcut", "cutoff" = "above_cut")
vd_downlab <- tibble(vd_downcut, below_cut) %>% rename("gene_id" = "vd_downcut", "cutoff" = "below_cut")

ap_labs <- bind_rows(ap_uplab, ap_downlab)
ml_labs <- bind_rows(ml_uplab, ml_downlab)
vd_labs <- bind_rows(vd_uplab, vd_downlab)

ml_comp <- inner_join(ml_labs, m8s15_olfrs, by = "gene_id") %>% inner_join(m8s15_100_tb, by = "gene_id")
mlcompup <- ml_comp %>% filter(cutoff == "up")
mlcompdown <- ml_comp %>% filter(cutoff == "down")
vd_comp <- inner_join(vd_labs, m9s7_olfrs, by = "gene_id") %>% inner_join(m9s7_100_tb, by = "gene_id")
vdcompup <- vd_comp %>% filter(cutoff == "up")
vdcompdown <- vd_comp %>% filter(cutoff == "down")
ap_comp <- inner_join(ap_labs, m10s18_olfrs, by = "gene_id") %>% inner_join(m10s18_100_tb, by = "gene_id")
apcompup <- ap_comp %>% filter(cutoff == "up")
apcompdown <- ap_comp %>% filter(cutoff == "down")

a <- cor(apcompup$m10s18_10, apcompup$tpm, method = "spearman")
b <- cor(apcompdown$m10s18_10, apcompdown$tpm, method = "spearman")
c <- cor(mlcompup$m8s15_10, mlcompup$tpm, method = "spearman")
d <- cor(mlcompdown$m8s15_10, mlcompdown$tpm, method = "spearman")
e <- cor(vdcompup$m9s7_10, vdcompup$tpm, method = "spearman")
f <- cor(vdcompdown$m9s7_10, vdcompdown$tpm, method = "spearman")

correlation <- c(a,b,c,d,e,f)
aflabs <- c("AP-above", "AP-below", "ML-above", "ML-below", "VD-above", "VD-below")
aftb <- tibble(correlation, aflabs)
ggplot(aftb) + geom_bar(aes(aflabs, correlation), stat = "identity")
```

#testing mapping of UTRs vs CDS for gencode v25
```{r}
m8s15_utr <- read_tsv("~/Downloads/gen25_noCDS/m8s15.genes.results")
m9s7_utr <- read_tsv("~/Desktop/obmap/heatmaps_starrsem/obmap_heatmaps_gen25_v200818/utronly.m9s7.genes.results")
m10s18_utr <- read_tsv("~/Desktop/obmap/heatmaps_starrsem/obmap_heatmaps_gen25_v200818/utronly.m10s18.genes.results")

m8s15_cds <- read_tsv("~/Desktop/obmap/heatmaps_starrsem/obmap_heatmaps_gen25_v200818/utrcds.M8S15.genes.results")
m9s7_cds <- read_tsv("~/Desktop/obmap/heatmaps_starrsem/obmap_heatmaps_gen25_v200818/utrcds.M9S07.genes.results")
m10s18_cds <- read_tsv("~/Desktop/obmap/heatmaps_starrsem/obmap_heatmaps_gen25_v200818/utrcds.M10S18.genes.results")

enstrans <- read_csv("~/Desktop/4marnus/convert_ensID_geneName.csv") %>%
  select(ens_id, gene_name) %>%
  unique()

test <- m8s15_utr %>% 
  select(gene_id, expected_count, TPM) %>% 
  rename(m8_count = expected_count, m8_tpm = TPM) %>%
  rowwise() %>%
  mutate(ens_id = str_split(gene_id, "[.]")[[1]][1]) %>%
  ungroup() %>%
  select(-gene_id) %>%
  select(ens_id, everything()) %>%
  left_join(enstrans, by = "ens_id")

utronly <- m8s15_utr %>% 
  select(gene_id, expected_count, TPM) %>% 
  rename(m8_count = expected_count, m8_tpm = TPM) %>%
  mutate(m9_count = m9s7_utr$expected_count, m9_tpm = m9s7_utr$TPM,
         m10_count = m10s18_utr$expected_count, m10_tpm = m10s18_utr$TPM) %>%
  rowwise() %>%
  mutate(ens_id = str_split(gene_id, "[.]")[[1]][1]) %>%
  ungroup() %>%
  select(-gene_id) %>%
  select(ens_id, everything()) %>%
  left_join(enstrans, by = "ens_id")

cdsonly <- m8s15_utr %>% 
  select(gene_id, expected_count, TPM) %>% 
  rename(m8_count = expected_count, m8_tpm = TPM) %>%
  mutate(m9_count = m9s7_cds$expected_count, m9_tpm = m9s7_cds$TPM,
         m10_count = m10s18_cds$expected_count, m10_tpm = m10s18_cds$TPM) %>%
  rowwise() %>%
  mutate(ens_id = str_split(gene_id, "[.]")[[1]][1]) %>%
  ungroup() %>%
  select(-gene_id) %>%
  select(ens_id, everything())

utr_vs_cds <- left_join(utronly, cdsonly, by = "ens_id", suffix = c(".utr", ".cds")) %>%
  filter(str_detect(gene_name, "^Olfr")) %>%
  filter(!str_detect(gene_name, "ps"))

count8 <- cor(utr_vs_cds$m8_count.utr, utr_vs_cds$m8_count.cds, method = "spearman")
count9 <- cor(utr_vs_cds$m9_count.utr, utr_vs_cds$m9_count.cds, method = "spearman")
count10 <- cor(utr_vs_cds$m10_count.utr, utr_vs_cds$m10_count.cds, method = "spearman")
tpm8 <- cor(utr_vs_cds$m8_tpm.utr, utr_vs_cds$m8_tpm.cds, method = "spearman")
tpm9 <- cor(utr_vs_cds$m9_tpm.utr, utr_vs_cds$m9_tpm.cds, method = "spearman")
tpm10 <- cor(utr_vs_cds$m10_tpm.utr, utr_vs_cds$m10_tpm.cds, method = "spearman")
#with all genes included, all spearman correlations are 0.999
#with only Olfrs (including pseudos), correlations are ~0.9996
#with only intact Olfrs, correlations are ~0.9997

#super high point is Olfr287
ggplot(utr_vs_cds %>% filter(!gene_name == "Olfr287")) + geom_point(aes(m8_count.utr, m8_count.cds, alpha = 0.25))

#make wiggle names
#m8s15_utr %>% select(gene_id) %>% rowwise() %>% mutate(ens_id = str_split(gene_id, "[.]")[[1]][1]) %>% ungroup() %>% left_join(enstrans, by = "ens_id") %>% filter(str_detect(gene_name, "Olfr")) %>% write_csv("~/Desktop/ensIDs_forORs.csv") 
```
